<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>immunogrep_singlechain_cdr3_cluster &mdash; IGREP 1.0.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="IGREP 1.0.0 documentation" href="../index.html" />
    <link rel="up" title="Module code" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../index.html">IGREP 1.0.0 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for immunogrep_singlechain_cdr3_cluster</h1><div class="highlight"><pre>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">generic_dna</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">datetime</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">operator</span> <span class="kn">import</span> <span class="n">itemgetter</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">gc</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">deque</span>
<span class="kn">import</span> <span class="nn">sys</span> 
<span class="kn">import</span> <span class="nn">math</span>

<span class="c">#for making folders in igrep</span>
<span class="kn">import</span> <span class="nn">immunogrep_file_system_tools</span> <span class="kn">as</span> <span class="nn">filesystem</span>
<span class="c">#for reading files</span>
<span class="kn">import</span> <span class="nn">immunogrep_read_file</span> <span class="kn">as</span> <span class="nn">readfile</span> 

<span class="c">#default dictionaries</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="c">#for printing pretty</span>
<span class="kn">from</span> <span class="nn">pprint</span> <span class="kn">import</span> <span class="n">pprint</span>

<span class="kn">from</span> <span class="nn">immunogrep_global_variables</span> <span class="kn">import</span> <span class="n">translation_var</span> <span class="c">#this key will signify the translation/translator key </span>
<span class="kn">from</span> <span class="nn">immunogrep_global_variables</span> <span class="kn">import</span> <span class="n">descriptor_symbol</span> <span class="c">#this key will signify the translation/translator key </span>
<span class="kn">from</span> <span class="nn">immunogrep_global_variables</span> <span class="kn">import</span> <span class="n">idIdentifier</span> <span class="c">#variable defining what SEQ_ID variable is in database </span>
<span class="kn">from</span> <span class="nn">immunogrep_global_variables</span> <span class="kn">import</span> <span class="n">fasta_file_delimiter</span>
<span class="c">#import immunogrep_useful_immunogrep_functions as useful</span>
<span class="kn">import</span> <span class="nn">immunogrep_useful_functions</span> <span class="kn">as</span> <span class="nn">useful</span>

<span class="kn">from</span> <span class="nn">random</span> <span class="kn">import</span> <span class="n">randint</span>

<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="c">#VERSION 4.0 applied on Version 7</span>
<span class="n">pairing_version</span> <span class="o">=</span> <span class="s">&#39;Pairing_v4.1&#39;</span>
<span class="c">#updates =&gt; allow isotyping </span>
	<span class="c">#=&gt; store SHM as mean SHM and variance </span>
	<span class="c">#=&gt; allow diversity calculations </span>
<span class="c">#updates 4.0 =&gt; produce a pairing &#39;annotation&#39; file </span>
	<span class="c">#the annotation file will store the &#39;paired-id&#39; of successful pairs , and their corresponding &#39;paired cluster id&#39;</span>
<span class="c">#updates 4.1 =&gt; </span>
	<span class="c">#use &#39;MODE&#39; for VGENE usage call:</span>
		<span class="c">#ONLY CONSIDER VGENES WHOSE SCORES ARE EQUAL TO THE TOP SCORE!!!</span>
		<span class="c">#RULE/ASSUMPTION: IF THERE IS NOT VGENE SCORE FIELD OR THE VGENE SCORE FIELD LENGTH != THE LENGHT OF VGENES THEN ONLY CONSIDER ALL VGENES PROVIDED</span>
	<span class="c">#we will no longer filter out sequences whose full length nt or aa field are empty </span>
	<span class="c">#add a new productivity rule =&gt; productive if there is no stop codon in the CDR3 sequence ONLY. This will become the default productivity rule  </span>
	

<span class="k">def</span> <span class="nf">get_ab_field_loc</span><span class="p">():</span>
	<span class="c">#structure of array in codeDict and collapsedict </span>
	<span class="c">#every R1-R2 read (that passes filters) will be stored as an array in the variable codeDict</span>
	<span class="c">#ONLY the first occurrence of H or L data from a pair will be stored in codeDict. </span>
	<span class="c">#the location of each data stored for either the heavy or light chain will be mapped by the following variable</span>
	<span class="c">#for example, lets if the Heavy chain is encountered first. It will be stored in codeDict as an array based on the indexes below. </span>
				 <span class="c">#the moment its Light chain pair is encountered, then the heavy chain data will be removed from the codeDict var and </span>
				 <span class="c">#both heavy and light chain data will be stored in collapsed_dict as a 2-D array where index 0 is heavy chain and index 1 is light chain . The third element of collapsed dict stores counts</span>
				 <span class="c">#-&gt; collapsedict[fullcode] = [heavy chain data array/list, light chain data array/list, COUNTS!!]			 </span>

	<span class="n">ab_field_loc</span> <span class="o">=</span> <span class="p">{</span>
		<span class="s">&#39;CDR3_SEQ&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="c">#store the cdr3 sequence in the first element of the array </span>
		<span class="s">&#39;VGENE&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span>
		<span class="s">&#39;DGENE&#39;</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span>
		<span class="s">&#39;JGENE&#39;</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span>
		<span class="s">&#39;LOCUS&#39;</span><span class="p">:</span><span class="mi">4</span><span class="p">,</span>
		<span class="s">&#39;MUT&#39;</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span>
		<span class="s">&#39;CDR3_LEN&#39;</span><span class="p">:</span><span class="mi">6</span><span class="p">,</span>
		<span class="s">&#39;AB_SEQ&#39;</span><span class="p">:</span><span class="mi">7</span><span class="p">,</span>
		<span class="s">&#39;CHAIN_CALL&#39;</span><span class="p">:</span><span class="mi">8</span><span class="p">,</span>
		<span class="s">&#39;ISOTYPE&#39;</span><span class="p">:</span><span class="mi">9</span><span class="p">,</span>
		<span class="s">&#39;VSCORES&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span>
		<span class="s">&#39;JSCORES&#39;</span><span class="p">:</span><span class="mi">11</span><span class="p">,</span>
		<span class="s">&#39;DSCORES&#39;</span><span class="p">:</span><span class="mi">12</span>
	<span class="p">}</span>
	<span class="k">return</span> <span class="n">ab_field_loc</span>

<span class="k">def</span> <span class="nf">get_required_field_names</span><span class="p">():</span>
	<span class="n">required_field_names</span> <span class="o">=</span> <span class="p">{</span>
			<span class="s">&#39;vgene&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;jgene&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;dgene&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;vgene_scores&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;jgene_scores&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;dgene_scores&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;cdr3_nt&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;seq_header&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;shm&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;full_len_ab_aa&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;full_len_ab_nt&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;functionality&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;recomb&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span>
		<span class="p">}</span>		
	<span class="k">return</span> <span class="n">required_field_names</span>

<span class="c">#the array for each heavy/light chain will always be this long</span>
<span class="n">num_elem_stored</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">v</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="n">get_ab_field_loc</span><span class="p">()</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()])</span><span class="o">+</span><span class="mi">1</span>  
<span class="n">allowed_file_formats</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;TAB&#39;</span><span class="p">,</span><span class="s">&#39;CSV&#39;</span><span class="p">,</span><span class="s">&#39;JSON&#39;</span><span class="p">,</span><span class="s">&#39;IMGT&#39;</span><span class="p">]</span>

<span class="c">#We will need to update the database with the results from IgFFT.  In order</span>
<span class="c">#to update teh database, we need a translator, so that we know what fields go where in the database</span>
<span class="k">def</span> <span class="nf">DatabaseTranslator</span><span class="p">(</span><span class="n">input_dictionary</span> <span class="o">=</span> <span class="p">{}):</span>
	<span class="n">key</span> <span class="o">=</span> <span class="n">translation_var</span>
	
	<span class="n">translator</span> <span class="o">=</span> <span class="p">{</span>								
			<span class="s">&quot;RECOMBINATION_FIELD&quot;</span><span class="p">:{</span> <span class="c">#THIS TELLS THE PROGRAM HOW TO DETERMINE WHETHER AN ANALYSIS/QUERY RESULT (from file) IS VDJ OR VJ</span>
					<span class="s">&quot;FIELD_NAME&quot;</span><span class="p">:</span> <span class="s">&quot;Recombination Type&quot;</span><span class="p">,</span> <span class="c">#name of the field in the file that will give information regarding the recombination type (VDJ OR VJ)					</span>
					<span class="s">&quot;EXPLICIT&quot;</span><span class="p">:</span><span class="bp">True</span><span class="p">,</span>
			<span class="p">},</span>
			<span class="s">&quot;FIELDS_UPDATE&quot;</span><span class="p">:{</span> 
				<span class="c">#key = field name  in database </span>
				<span class="c">#value = field name in file</span>
				<span class="c">#this will map all of the fields in the file to the proper location in the database. I.E. If I list VGENES as the column name/field name, then i want to map VREGION.VGENES:VGENES (because VREGION.VGENES is the name in the database)							</span>
				<span class="n">idIdentifier</span><span class="p">:</span><span class="n">idIdentifier</span><span class="p">,</span>
				<span class="s">&quot;SEQUENCE&quot;</span><span class="p">:</span><span class="s">&quot;Sequence&quot;</span><span class="p">,</span>				
				<span class="s">&quot;SEQUENCE_HEADER&quot;</span><span class="p">:</span><span class="s">&quot;Header&quot;</span><span class="p">,</span>								
				<span class="s">&quot;PAIRING.PAIRED_ID&quot;</span><span class="p">:</span><span class="s">&#39;PairedID&#39;</span><span class="p">,</span>
				<span class="s">&quot;PAIRING.PAIRED_CLUSTER&quot;</span><span class="p">:</span><span class="s">&#39;PairedClusterId&#39;</span><span class="p">,</span>
				<span class="s">&quot;PAIRING.CONFIDENCE&quot;</span><span class="p">:</span><span class="s">&#39;Confidence&#39;</span><span class="p">,</span>
				<span class="s">&quot;PAIRING.DOMINANCE&quot;</span><span class="p">:</span><span class="s">&#39;Dominance&#39;</span><span class="p">,</span>
				<span class="s">&#39;COMMAND&#39;</span><span class="p">:</span><span class="s">&#39;Command&#39;</span>
			<span class="p">}</span>
	<span class="p">}</span>
	
	<span class="n">input_dictionary</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">translator</span>		
	
	<span class="k">return</span> <span class="n">input_dictionary</span>

<span class="k">def</span> <span class="nf">FloatRange</span><span class="p">(</span><span class="n">minv</span><span class="p">,</span><span class="n">maxv</span><span class="p">,</span><span class="n">step</span><span class="p">,</span><span class="n">num_dec</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
	<span class="k">if</span> <span class="n">step</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
		<span class="n">step</span> <span class="o">=</span><span class="mf">0.01</span>
	<span class="n">minv</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">float</span><span class="p">(</span><span class="n">minv</span><span class="p">))</span>
	<span class="n">maxv</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">float</span><span class="p">(</span><span class="n">maxv</span><span class="p">))</span>
	
	<span class="n">x</span> <span class="o">=</span> <span class="n">minv</span>
	<span class="k">if</span> <span class="n">maxv</span><span class="o">&lt;</span><span class="n">minv</span><span class="p">:</span>
		<span class="n">min2</span><span class="o">=</span><span class="n">minv</span>
		<span class="n">minv</span><span class="o">=</span><span class="n">maxv</span>
		<span class="n">maxv</span><span class="o">=</span><span class="n">min2</span>
	<span class="k">if</span> <span class="n">minv</span><span class="o">==</span><span class="n">maxv</span><span class="p">:</span>
		<span class="k">yield</span> <span class="n">minv</span>
		<span class="n">x</span><span class="o">+=</span><span class="mi">1</span>
		<span class="n">minv</span><span class="o">+=</span><span class="mi">1</span>
		<span class="c">#yield min+1</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">while</span> <span class="n">x</span><span class="o">&lt;=</span><span class="n">maxv</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">num_dec</span><span class="p">:</span>
				<span class="k">yield</span> <span class="nb">round</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">num_dec</span><span class="p">)</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="k">yield</span> <span class="n">x</span>
			<span class="n">x</span><span class="o">+=</span><span class="nb">float</span><span class="p">(</span><span class="n">step</span><span class="p">)</span>
	<span class="k">yield</span> <span class="n">maxv</span>

<span class="c">#use output from IMGT program to determine productivity</span>
<span class="k">def</span> <span class="nf">IMGTProductivity</span><span class="p">(</span><span class="n">field_row</span><span class="p">):</span>
	<span class="k">global</span> <span class="n">functionality_field</span>
	<span class="k">if</span> <span class="n">field_row</span><span class="p">[</span><span class="n">functionality_field</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;unproductive&#39;</span> <span class="ow">or</span> <span class="n">field_row</span><span class="p">[</span><span class="n">functionality_field</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;No results&#39;</span> <span class="ow">or</span> <span class="ow">not</span><span class="p">(</span><span class="n">field_row</span><span class="p">[</span><span class="n">functionality_field</span><span class="p">]):</span>
		<span class="k">return</span> <span class="bp">False</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">return</span> <span class="bp">True</span>

<span class="c">#use output from IGBLAST program to determine productivity</span>
<span class="c">#(have not handled this function yet)</span>
<span class="k">def</span> <span class="nf">IGBLASTProductivity</span><span class="p">(</span><span class="n">field_row</span><span class="p">):</span>
	<span class="k">return</span> <span class="bp">True</span> 

<span class="c">#use output from IGFFT program to determine productivity</span>
<span class="k">def</span> <span class="nf">IGFFTProductivity</span><span class="p">(</span><span class="n">field_row</span><span class="p">):</span>
	<span class="k">global</span> <span class="n">functionality_field</span>
	<span class="k">if</span> <span class="n">field_row</span><span class="p">[</span><span class="n">functionality_field</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">==</span> <span class="s">&#39;NO&#39;</span> <span class="ow">or</span> <span class="ow">not</span><span class="p">(</span><span class="n">field_row</span><span class="p">[</span><span class="n">functionality_field</span><span class="p">]):</span>
		<span class="k">return</span> <span class="bp">False</span>
	<span class="k">else</span><span class="p">:</span>		
		<span class="k">return</span> <span class="bp">True</span>

<span class="c">#use output from MIXCR program to determine productivity</span>
<span class="k">def</span> <span class="nf">MIXCRProductivity</span><span class="p">(</span><span class="n">field_row</span><span class="p">):</span>	
	<span class="k">global</span> <span class="n">full_aa_field</span>
	<span class="k">if</span> <span class="s">&#39;*&#39;</span> <span class="ow">in</span> <span class="n">field_row</span><span class="p">[</span><span class="n">full_aa_field</span><span class="p">]</span> <span class="ow">or</span> <span class="s">&#39;_&#39;</span> <span class="ow">in</span> <span class="n">field_row</span><span class="p">[</span><span class="n">full_aa_field</span><span class="p">]:</span>
		<span class="k">return</span> <span class="bp">False</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">return</span> <span class="bp">True</span>


<span class="c">#general function for determining if a sequence is productive or not </span>
<div class="viewcode-block" id="GeneralProductivity"><a class="viewcode-back" href="../VHVLPairing.html#immunogrep_singlechain_cdr3_cluster.GeneralProductivity">[docs]</a><span class="k">def</span> <span class="nf">GeneralProductivity</span><span class="p">(</span><span class="n">field_row</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		General function for determining if an entire Antibody amino acid sequence is productive or not </span>
<span class="sd">	</span>
<span class="sd">		Function will return True or False based on rule </span>
<span class="sd">		Input variable field_row: this will represent all antibody annotation field for a specific sequence the provided annotation file</span>
<span class="sd">		</span>
<span class="sd">		General steps:</span>
<span class="sd">			1) Extract the field referring to the amino acid sequence of the antibody 			</span>
<span class="sd">			4) If there is a stop codon (*) in sequence </span>
<span class="sd">				=&gt; return false </span>
<span class="sd">			5) return true</span>
<span class="sd">	</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="k">global</span> <span class="n">full_aa_field</span>
	<span class="k">if</span> <span class="s">&#39;*&#39;</span> <span class="ow">in</span> <span class="n">field_row</span><span class="p">[</span><span class="n">full_aa_field</span><span class="p">]:</span>
		<span class="k">return</span> <span class="bp">False</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">return</span> <span class="bp">True</span>
</div>
<div class="viewcode-block" id="CDR3Productivity"><a class="viewcode-back" href="../VHVLPairing.html#immunogrep_singlechain_cdr3_cluster.CDR3Productivity">[docs]</a><span class="k">def</span> <span class="nf">CDR3Productivity</span><span class="p">(</span><span class="n">field_row</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	</span>
<span class="sd">		Rule for CDR3 productivity. 		</span>
<span class="sd">		</span>
<span class="sd">		Function will return True or False based on rule </span>
<span class="sd">		Input variable field_row: this will represent all antibody annotation field for a specific sequence the provided annotation file</span>
<span class="sd">		</span>
<span class="sd">		General steps:</span>
<span class="sd">			1) Extract the CDR3 nucleotide field </span>
<span class="sd">			2) If length of CDR3 is not a multiple of 3 and length of cdr3 &lt; 9</span>
<span class="sd">				=&gt; return false </span>
<span class="sd">			3) Translate CDR3 to amino acid </span>
<span class="sd">			4) If there is a stop codon (*) in sequence</span>
<span class="sd">				=&gt; return false </span>
<span class="sd">			5) return true</span>
<span class="sd">		</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="k">global</span> <span class="n">cdr3_field</span>
	<span class="n">cdr3_nt</span> <span class="o">=</span> <span class="n">field_row</span><span class="p">[</span><span class="n">cdr3_field</span><span class="p">]</span>
	<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cdr3_nt</span><span class="p">)</span><span class="o">%</span><span class="mi">3</span><span class="o">!=</span><span class="mi">0</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">cdr3_nt</span><span class="p">)</span><span class="o">&lt;</span><span class="mi">9</span><span class="p">:</span>
		<span class="k">return</span> <span class="bp">False</span>		
	<span class="n">cdr3_aa</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">cdr3_nt</span><span class="p">,</span><span class="n">generic_dna</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
	<span class="k">if</span> <span class="s">&#39;*&#39;</span> <span class="ow">in</span> <span class="n">cdr3_aa</span><span class="p">:</span>
		<span class="k">return</span> <span class="bp">False</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">return</span> <span class="bp">True</span>

	</div>
<span class="k">def</span> <span class="nf">populate_clusters</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span><span class="n">delimiter</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">):</span>
	<span class="n">d</span> <span class="o">=</span> <span class="p">{}</span>
	<span class="k">if</span> <span class="ow">not</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename</span><span class="p">)):</span>
		<span class="k">return</span> <span class="n">d</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_in</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">line_text</span> <span class="ow">in</span> <span class="n">f_in</span><span class="p">:</span>			
			<span class="n">line_text</span> <span class="o">=</span> <span class="n">line_text</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
			<span class="n">line</span> <span class="o">=</span> <span class="n">line_text</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">delimiter</span><span class="p">)</span>
		
			<span class="n">SorH</span><span class="o">=</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="c">#Seed or Hit</span>
			<span class="n">number</span><span class="o">=</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="c">#seed number</span>
			
			<span class="k">if</span> <span class="s">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">[</span><span class="mi">8</span><span class="p">]:</span>
				<span class="n">info</span><span class="o">=</span><span class="n">line</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">info</span><span class="o">=</span><span class="n">line</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
			<span class="n">counts</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">info</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
			
			<span class="c">#identify seeds (S) vs hits (H)</span>
			<span class="k">if</span> <span class="n">SorH</span><span class="o">==</span><span class="s">&#39;S&#39;</span><span class="p">:</span>
				<span class="n">d</span><span class="p">[</span><span class="n">number</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">counts</span><span class="p">,]</span>
				<span class="n">d</span><span class="p">[</span><span class="n">number</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">info</span><span class="p">)</span>
			<span class="k">if</span> <span class="n">SorH</span><span class="o">==</span><span class="s">&#39;H&#39;</span><span class="p">:</span>
				<span class="n">d</span><span class="p">[</span><span class="n">number</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">+=</span><span class="n">counts</span>
				<span class="n">d</span><span class="p">[</span><span class="n">number</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">info</span><span class="p">)</span>		
	<span class="k">return</span> <span class="n">d</span> 

<span class="k">def</span> <span class="nf">gene_hist</span><span class="p">(</span><span class="n">d</span><span class="p">,</span><span class="n">filename</span><span class="p">,</span><span class="n">geneCall</span><span class="p">):</span>
	
	<span class="n">summary</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span><span class="s">&#39;a&#39;</span><span class="p">)</span>
	
	<span class="n">hist</span><span class="o">=</span><span class="nb">dict</span><span class="p">()</span>
	
	<span class="k">if</span> <span class="n">geneCall</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
		<span class="n">Title</span><span class="o">=</span><span class="s">&#39;VH gene analysis&#39;</span>
		<span class="n">genePos</span><span class="o">=</span><span class="mi">5</span>
	<span class="k">if</span> <span class="n">geneCall</span><span class="o">==</span><span class="mi">2</span><span class="p">:</span>
		<span class="n">Title</span><span class="o">=</span><span class="s">&#39;DH gene analysis&#39;</span>
		<span class="n">genePos</span><span class="o">=</span><span class="mi">6</span>
	<span class="k">if</span> <span class="n">geneCall</span><span class="o">==</span><span class="mi">3</span><span class="p">:</span>
		<span class="n">Title</span><span class="o">=</span><span class="s">&#39;JH gene analysis&#39;</span>
		<span class="n">genePos</span><span class="o">=</span><span class="mi">7</span>
	<span class="k">if</span> <span class="n">geneCall</span><span class="o">==</span><span class="mi">4</span><span class="p">:</span>
		<span class="n">Title</span><span class="o">=</span><span class="s">&#39;VL gene analysis&#39;</span>
		<span class="n">genePos</span><span class="o">=</span><span class="mi">8</span>
	<span class="k">if</span> <span class="n">geneCall</span><span class="o">==</span><span class="mi">5</span><span class="p">:</span>
		<span class="n">Title</span><span class="o">=</span><span class="s">&#39;JL gene analysis&#39;</span>
		<span class="n">genePos</span><span class="o">=</span><span class="mi">9</span>
	
	<span class="n">total</span><span class="o">=</span><span class="mi">0</span>
	<span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">d</span><span class="p">:</span>
		<span class="n">gene</span><span class="o">=</span><span class="n">d</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">)[</span><span class="n">genePos</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;*&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
		<span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">hist</span><span class="p">:</span>
			<span class="n">hist</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">hist</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">=</span><span class="mi">1</span>
		<span class="n">total</span><span class="o">+=</span><span class="mi">1</span>	

	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n\n</span><span class="s">******  </span><span class="si">%s</span><span class="s">  ******</span><span class="se">\n\n</span><span class="s">&#39;</span> <span class="o">%</span><span class="p">(</span><span class="n">Title</span><span class="p">))</span>
	<span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">hist</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">hist</span><span class="p">[</span><span class="n">i</span><span class="p">]),</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
		<span class="n">percent</span><span class="o">=</span><span class="nb">round</span><span class="p">(</span><span class="mf">100.0</span><span class="o">*</span><span class="n">hist</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">/</span><span class="n">total</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
		<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{:&lt;12} {:&lt;12} {:&lt;12}</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="p">,</span><span class="n">hist</span><span class="p">[</span><span class="n">key</span><span class="p">],</span><span class="n">percent</span><span class="p">))</span>	

	
<span class="c">#variable for determining whether the antibody is heavy or light chain </span>
<span class="n">chain_call</span> <span class="o">=</span> <span class="p">{</span>
	<span class="s">&#39;VDJ&#39;</span><span class="p">:[</span><span class="s">&#39;TRB&#39;</span><span class="p">,</span><span class="s">&#39;TB&#39;</span><span class="p">,</span><span class="s">&#39;IGH&#39;</span><span class="p">,</span><span class="s">&#39;TRD&#39;</span><span class="p">],</span>
	<span class="s">&#39;VJ&#39;</span><span class="p">:[</span><span class="s">&#39;IGK&#39;</span><span class="p">,</span><span class="s">&#39;TRA&#39;</span><span class="p">,</span><span class="s">&#39;TA&#39;</span><span class="p">,</span><span class="s">&#39;IGL&#39;</span><span class="p">,</span><span class="s">&#39;TRG&#39;</span><span class="p">],</span>	
<span class="p">}</span>


<span class="c">#this is a simple variable for mapping the fields we require to run the analysis to the field names in different annotation files </span>
<span class="c">#so when reading files we know that the field name &#39;Functionality_1&#39; in IMGT files correponds to our functionality variable we use in the analysis </span>
<span class="n">pairing_settings</span> <span class="o">=</span> <span class="p">{</span>
	<span class="s">&#39;IMGT&#39;</span><span class="p">:{</span>
		<span class="s">&#39;fields_for_analysis&#39;</span><span class="p">:{</span>
			<span class="s">&#39;functionality&#39;</span><span class="p">:</span><span class="s">&#39;Functionality_1&#39;</span><span class="p">,</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&#39;Sequence ID_1&#39;</span><span class="p">,</span>
			<span class="s">&#39;vgene&#39;</span><span class="p">:</span><span class="s">&#39;V-GENE and allele_1&#39;</span><span class="p">,</span>
			<span class="s">&#39;jgene&#39;</span><span class="p">:</span><span class="s">&#39;J-GENE and allele_1&#39;</span><span class="p">,</span>
			<span class="s">&#39;dgene&#39;</span><span class="p">:</span><span class="s">&#39;D-GENE and allele_1&#39;</span><span class="p">,</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&#39;Sequence_1&#39;</span><span class="p">,</span>
			<span class="s">&#39;cdr3_nt&#39;</span><span class="p">:</span><span class="s">&#39;CDR3-IMGT_3&#39;</span><span class="p">,</span>
			<span class="s">&#39;seq_header&#39;</span><span class="p">:</span><span class="s">&#39;Sequence ID_1&#39;</span><span class="p">,</span>
			<span class="s">&#39;shm&#39;</span><span class="p">:</span><span class="s">&#39;VREGION.SHM.NT_PER&#39;</span><span class="p">,</span>			
			<span class="s">&#39;full_len_ab_aa&#39;</span><span class="p">:</span><span class="s">&#39;PREDICTED_AB_SEQ.AA&#39;</span><span class="p">,</span><span class="c">#THIS FIELD IS NOT PRESENT IN THE IMGT FILE, BUT IT DOES GET ADDED TO THE INFORMATION WHEN READING IMGT FILES USING OUR READER CLASS			</span>
			<span class="s">&#39;full_len_ab_nt&#39;</span><span class="p">:</span><span class="s">&#39;PREDICTED_AB_SEQ.NT&#39;</span><span class="p">,</span><span class="c">#THIS FIELD IS NOT PRESENT IN THE IMGT FILE, BUT IT DOES GET ADDED TO THE INFORMATION WHEN READING IMGT FILES USING OUR READER CLASS			</span>
			<span class="s">&#39;vgene_scores&#39;</span><span class="p">:</span><span class="s">&#39;V-REGION score_1&#39;</span><span class="p">,</span>
			<span class="s">&#39;dgene_scores&#39;</span><span class="p">:</span><span class="s">&#39;D-REGION score_1&#39;</span><span class="p">,</span>
			<span class="s">&#39;jgene_scores&#39;</span><span class="p">:</span><span class="s">&#39;J-REGION score_1&#39;</span><span class="p">,</span>
			<span class="s">&#39;recomb&#39;</span><span class="p">:</span><span class="s">&#39;R_TYPE&#39;</span>
			
		<span class="p">},</span>
		<span class="s">&#39;productivity_function&#39;</span><span class="p">:</span><span class="n">IMGTProductivity</span> <span class="c">#when handling IMGT anaysis, use the function described above (IMGTProductivity) to determine whether a sequence is productive or not </span>
	<span class="p">},</span>
	<span class="s">&#39;IGBLAST&#39;</span><span class="p">:{</span>
		<span class="s">&#39;fields_for_analysis&#39;</span><span class="p">:{</span>
			<span class="s">&#39;functionality&#39;</span><span class="p">:</span><span class="s">&#39;PRODUCTIVE&#39;</span><span class="p">,</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&quot;FULL_SEQ&quot;</span><span class="p">,</span>
			<span class="s">&#39;vgene&#39;</span><span class="p">:</span><span class="s">&#39;VREGION.VGENES&#39;</span><span class="p">,</span>
			<span class="s">&#39;jgene&#39;</span><span class="p">:</span><span class="s">&#39;JREGION.JGENES&#39;</span><span class="p">,</span>
			<span class="s">&#39;dgene&#39;</span><span class="p">:</span><span class="s">&#39;DREGION.DGENES&#39;</span><span class="p">,</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&#39;FULL_SEQ&#39;</span><span class="p">,</span>
			<span class="s">&#39;cdr3_nt&#39;</span><span class="p">:</span><span class="s">&#39;CDR3.NT&#39;</span><span class="p">,</span>
			<span class="s">&#39;seq_header&#39;</span><span class="p">:</span><span class="s">&#39;DOCUMENTHEADER&#39;</span><span class="p">,</span>
			<span class="s">&#39;shm&#39;</span><span class="p">:</span><span class="s">&quot;VREGION.SHM_NT_PER&quot;</span><span class="p">,</span>			
			<span class="s">&#39;full_len_ab_aa&#39;</span><span class="p">:</span><span class="s">&#39;PREDICTED_AB_SEQ.AA&#39;</span><span class="p">,</span><span class="c">#optional field but preferred</span>
			<span class="s">&#39;full_len_ab_nt&#39;</span><span class="p">:</span><span class="s">&#39;PREDICTED_AB_SEQ.NT&#39;</span><span class="p">,</span>			
			<span class="s">&#39;vgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;VREGION.VGENE_SCORE&quot;</span><span class="p">,</span>
			<span class="s">&#39;jgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;JREGION.VGENE_SCORE&quot;</span><span class="p">,</span>
			<span class="s">&#39;dgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;DREGION.VGENE_SCORE&quot;</span><span class="p">,</span>
			<span class="s">&#39;recomb&#39;</span><span class="p">:</span><span class="s">&#39;RECOMBINATION_TYPE&#39;</span>
		<span class="p">},</span>                    
		<span class="s">&#39;productivity_function&#39;</span><span class="p">:</span><span class="n">CDR3Productivity</span>
	<span class="p">},</span>
	<span class="s">&#39;IGFFT&#39;</span><span class="p">:{</span>
		<span class="s">&#39;fields_for_analysis&#39;</span><span class="p">:{</span>
			<span class="s">&#39;functionality&#39;</span><span class="p">:</span><span class="s">&#39;Productive&#39;</span><span class="p">,</span>
			<span class="s">&#39;vgene&#39;</span><span class="p">:</span><span class="s">&#39;Top_V-Gene_Hits&#39;</span><span class="p">,</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&#39;Sequence&#39;</span><span class="p">,</span>
			<span class="s">&#39;jgene&#39;</span><span class="p">:</span><span class="s">&#39;Top_J-Gene_Hits&#39;</span><span class="p">,</span>
			<span class="s">&#39;dgene&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span><span class="c">#Top_D-Gene_Hits</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&#39;Sequence&#39;</span><span class="p">,</span>
			<span class="s">&#39;cdr3_nt&#39;</span><span class="p">:</span><span class="s">&#39;CDR3_Sequence.NT&#39;</span><span class="p">,</span>
			<span class="s">&#39;seq_header&#39;</span><span class="p">:</span><span class="s">&#39;Header&#39;</span><span class="p">,</span>
			<span class="s">&#39;shm&#39;</span><span class="p">:</span><span class="s">&#39;VRegion.SHM.Per_nt&#39;</span><span class="p">,</span><span class="c">#&#39;VRegion.SHM.NT&#39;,</span>
			<span class="s">&#39;full_len_ab_aa&#39;</span><span class="p">:</span><span class="s">&#39;Full_Length_Sequence.AA&#39;</span><span class="p">,</span><span class="c">#optional field but preferred</span>
			<span class="s">&#39;full_len_ab_nt&#39;</span><span class="p">:</span><span class="s">&#39;Full_Length_Sequence.NT&#39;</span><span class="p">,</span>
			<span class="s">&#39;isotype&#39;</span><span class="p">:</span><span class="s">&#39;Isotype&#39;</span><span class="p">,</span>
			<span class="s">&#39;vgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;V-Gene_Alignment_Scores&quot;</span><span class="p">,</span>
			<span class="s">&#39;jgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;J-Gene_Alignment_Scores&quot;</span><span class="p">,</span>
			<span class="s">&#39;dgene_scores&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span>
			<span class="s">&#39;recomb&#39;</span><span class="p">:</span><span class="s">&#39;Recombination_Type&#39;</span>
		<span class="p">},</span>
		<span class="s">&#39;productivity_function&#39;</span><span class="p">:</span><span class="n">CDR3Productivity</span>
	<span class="p">},</span>
	<span class="s">&#39;MIXCR&#39;</span><span class="p">:{</span>
		<span class="s">&#39;fields_for_analysis&#39;</span><span class="p">:{</span>
			<span class="s">&#39;functionality&#39;</span><span class="p">:</span><span class="s">&#39;Productivity&#39;</span><span class="p">,</span>
			<span class="s">&#39;vgene&#39;</span><span class="p">:</span><span class="s">&quot;All V hits&quot;</span><span class="p">,</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&#39;Sequence&#39;</span><span class="p">,</span>
			<span class="s">&#39;jgene&#39;</span><span class="p">:</span><span class="s">&quot;All J hits&quot;</span><span class="p">,</span>
			<span class="s">&#39;dgene&#39;</span><span class="p">:</span><span class="s">&quot;All D hits&quot;</span><span class="p">,</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&#39;Sequence&#39;</span><span class="p">,</span>
			<span class="s">&#39;cdr3_nt&#39;</span><span class="p">:</span><span class="s">&#39;N. Seq. CDR3&#39;</span><span class="p">,</span>
			<span class="s">&#39;seq_header&#39;</span><span class="p">:</span><span class="s">&#39;Seqheader&#39;</span><span class="p">,</span>
			<span class="s">&#39;shm&#39;</span><span class="p">:</span> <span class="s">&#39;VGENE: Shm.per&#39;</span><span class="p">,</span><span class="c">#&#39;VGENE: Shm.nt&#39;,</span>
			<span class="s">&#39;full_len_ab_aa&#39;</span><span class="p">:</span><span class="s">&#39;Full AA&#39;</span><span class="p">,</span><span class="c">#optional field but preferred</span>
			<span class="s">&#39;full_len_ab_nt&#39;</span><span class="p">:</span><span class="s">&#39;Full NT&#39;</span><span class="p">,</span>
			<span class="s">&#39;isotype&#39;</span><span class="p">:</span><span class="s">&#39;All C hits&#39;</span><span class="p">,</span>
			<span class="s">&#39;vgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;All V scores&quot;</span><span class="p">,</span>
			<span class="s">&#39;jgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;All J scores&quot;</span><span class="p">,</span>
			<span class="s">&#39;dgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;All D scores&quot;</span><span class="p">,</span>
			<span class="s">&#39;recomb&#39;</span><span class="p">:</span><span class="s">&#39;Recombination Type&#39;</span>
		<span class="p">},</span>
		<span class="s">&#39;productivity_function&#39;</span><span class="p">:</span><span class="n">CDR3Productivity</span>
	<span class="p">},</span>
	
	<span class="s">&#39;DATABASE&#39;</span><span class="p">:{</span>
		<span class="s">&#39;fields_for_analysis&#39;</span><span class="p">:{</span>
			<span class="s">&#39;functionality&#39;</span><span class="p">:</span><span class="s">&#39;PRODUCTIVE&#39;</span><span class="p">,</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&#39;SEQUENCE&#39;</span><span class="p">,</span>
			<span class="s">&#39;vgene&#39;</span><span class="p">:</span><span class="s">&#39;VREGION.VGENES&#39;</span><span class="p">,</span>
			<span class="s">&#39;jgene&#39;</span><span class="p">:</span><span class="s">&#39;JREGION.JGENES&#39;</span><span class="p">,</span>
			<span class="s">&#39;dgene&#39;</span><span class="p">:</span><span class="s">&#39;DREGION.DGENES&#39;</span><span class="p">,</span>
			<span class="s">&#39;raw_seq_nt&#39;</span><span class="p">:</span><span class="s">&#39;PREDICTED_AB_SEQ.NT&#39;</span><span class="p">,</span>
			<span class="s">&#39;cdr3_nt&#39;</span><span class="p">:</span><span class="s">&#39;CDR3.NT&#39;</span><span class="p">,</span>
			<span class="s">&#39;seq_header&#39;</span><span class="p">:</span><span class="s">&#39;SEQUENCE_HEADER&#39;</span><span class="p">,</span>
			<span class="s">&#39;shm&#39;</span><span class="p">:</span><span class="s">&#39;VREGION.SHM.NT_PER&#39;</span><span class="p">,</span>
			<span class="s">&#39;full_len_ab_aa&#39;</span><span class="p">:</span><span class="s">&#39;PREDICTED_AB_SEQ.AA&#39;</span><span class="p">,</span><span class="c">#optional field but preferred</span>
			<span class="s">&#39;full_len_ab_nt&#39;</span><span class="p">:</span><span class="s">&#39;PREDICTED_AB_SEQ.NT&#39;</span><span class="p">,</span>
			<span class="s">&#39;isotype&#39;</span><span class="p">:</span><span class="s">&#39;ISOTYPE.GENE&#39;</span><span class="p">,</span>
			<span class="s">&#39;vgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;VREGION.VGENE_SCORES&quot;</span><span class="p">,</span>
			<span class="s">&#39;jgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;JREGION.JGENE_SCORES&quot;</span><span class="p">,</span>
			<span class="s">&#39;dgene_scores&#39;</span><span class="p">:</span><span class="s">&quot;DREGION.DGENE_SCORES&quot;</span><span class="p">,</span>
			<span class="s">&#39;recomb&#39;</span><span class="p">:</span><span class="s">&#39;RECOMBINATION_TYPE&#39;</span>
		<span class="p">},</span>
		<span class="s">&#39;productivity_function&#39;</span><span class="p">:</span><span class="n">CDR3Productivity</span>
	<span class="p">}</span>
<span class="p">}</span>


<span class="n">supported_analyses</span> <span class="o">=</span> <span class="n">pairing_settings</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

<span class="k">def</span> <span class="nf">GetHeaderInfo</span><span class="p">(</span><span class="n">file_data</span><span class="p">,</span><span class="n">header_var</span><span class="p">):</span>
	<span class="n">h</span> <span class="o">=</span> <span class="n">file_data</span><span class="p">[</span><span class="n">header_var</span><span class="p">]</span>
	
	<span class="c">#check if ididnetifier is in the file first </span>
	<span class="k">if</span> <span class="n">idIdentifier</span> <span class="ow">in</span> <span class="n">file_data</span> <span class="ow">and</span> <span class="n">file_data</span><span class="p">[</span><span class="n">idIdentifier</span><span class="p">]:</span>
		<span class="nb">id</span> <span class="o">=</span> <span class="n">file_data</span><span class="p">[</span><span class="n">idIdentifier</span><span class="p">]</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="c">#its not in the file, so it might in header </span>
		<span class="nb">id</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
		<span class="k">if</span> <span class="n">fasta_file_delimiter</span> <span class="ow">in</span> <span class="n">h</span><span class="p">:</span>
			<span class="c">#check header </span>
			<span class="n">tmp</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">fasta_file_delimiter</span><span class="p">)</span>
			<span class="k">try</span><span class="p">:</span>
				<span class="n">additional_data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
			<span class="k">except</span><span class="p">:</span>
				<span class="n">additional_data</span><span class="o">=</span><span class="p">{}</span>
			<span class="k">if</span> <span class="n">idIdentifier</span> <span class="ow">in</span> <span class="n">additional_data</span><span class="p">:</span>
				<span class="nb">id</span> <span class="o">=</span> <span class="n">additional_data</span><span class="p">[</span><span class="n">idIdentifier</span><span class="p">]</span>
	<span class="k">return</span> <span class="p">[</span><span class="n">h</span><span class="p">,</span><span class="nb">id</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">initialize_input_files</span><span class="p">(</span><span class="n">analysis_method</span><span class="p">,</span><span class="n">list_of_filetypes</span><span class="p">,</span><span class="n">list_of_files</span><span class="p">):</span>
	<span class="k">global</span> <span class="n">annotation_headers</span>
	<span class="c">#load all of the filenames and open the files using our class for reading different files </span>
	<span class="k">if</span> <span class="n">list_of_filetypes</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span>
		<span class="n">list_of_file_reading</span> <span class="o">=</span> <span class="p">[</span><span class="n">readfile</span><span class="o">.</span><span class="n">immunogrepFile</span><span class="p">(</span><span class="nb">file</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="nb">file</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">list_of_files</span><span class="p">)]</span>
	<span class="k">elif</span> <span class="n">list_of_filetypes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span><span class="s">&#39;IMGT&#39;</span><span class="p">:</span>
		<span class="n">list_of_file_reading</span> <span class="o">=</span> <span class="p">[</span><span class="n">readfile</span><span class="o">.</span><span class="n">immunogrepFile</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span><span class="s">&#39;IMGT&#39;</span><span class="p">,</span><span class="n">required_files</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">8</span><span class="p">],</span><span class="n">field_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;Sequence_1&#39;</span><span class="p">,</span><span class="s">&#39;V-D-J-REGION_3&#39;</span><span class="p">,</span><span class="s">&#39;V-D-J-REGION_5&#39;</span><span class="p">,</span><span class="s">&#39;V-J-REGION_3&#39;</span><span class="p">,</span><span class="s">&#39;V-J-REGION_5&#39;</span><span class="p">,</span><span class="s">&#39;CDR3-IMGT_5&#39;</span><span class="p">,</span><span class="s">&#39;Functionality_1&#39;</span><span class="p">,</span><span class="s">&#39;V-GENE and allele_1&#39;</span><span class="p">,</span><span class="s">&#39;D-GENE and allele_1&#39;</span><span class="p">,</span><span class="s">&#39;J-GENE and allele_1&#39;</span><span class="p">,</span><span class="s">&#39;CDR3-IMGT_3&#39;</span><span class="p">,</span><span class="s">&#39;Sequence ID_1&#39;</span><span class="p">,</span><span class="s">&#39;V-REGION Nb of mutations_8&#39;</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="nb">file</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">list_of_files</span><span class="p">)]</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">list_of_file_reading</span> <span class="o">=</span> <span class="p">[</span><span class="n">readfile</span><span class="o">.</span><span class="n">immunogrepFile</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span><span class="n">list_of_filetypes</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="nb">file</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">list_of_files</span><span class="p">)]</span>
		
	<span class="k">return</span> <span class="n">list_of_file_reading</span>


<div class="viewcode-block" id="ProcessGene"><a class="viewcode-back" href="../VHVLPairing.html#immunogrep_singlechain_cdr3_cluster.ProcessGene">[docs]</a><span class="k">def</span> <span class="nf">ProcessGene</span><span class="p">(</span><span class="n">gene</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		Removes the allele calls from genes</span>
<span class="sd">		</span>
<span class="sd">		Assumption: </span>
<span class="sd">			</span>
<span class="sd">			Multiple genes are seperated by &#39;,&#39; </span>
<span class="sd">			</span>
<span class="sd">			Alleles are seperated by &#39;*&#39; </span>
<span class="sd">			</span>
<span class="sd">			If a gene is seperated by multiple spaces, then the gene should be identified by a gene that contains either - or &#39;*&#39; </span>
<span class="sd">			For example: </span>
<span class="sd">				Imgt genes may be: </span>
<span class="sd">					Homo sapiens IGHV1-3*01</span>
<span class="sd">					We only want to isolate the word IGHV1-3</span>
<span class="sd">	&quot;&quot;&quot;</span>
	
	<span class="k">if</span> <span class="ow">not</span> <span class="n">gene</span><span class="p">:</span>
		<span class="k">return</span> <span class="s">&#39;&#39;</span>	
	<span class="c">#extract all genes in field </span>
	<span class="n">gene_array</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)</span>
	<span class="k">for</span> <span class="n">gene_num</span><span class="p">,</span><span class="n">each_gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_array</span><span class="p">):</span>
		<span class="c">#split each gene by spaces. Go through each word</span>
		<span class="n">split_words</span>  <span class="o">=</span> <span class="n">each_gene</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39; &#39;</span><span class="p">)</span>
		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">split_words</span><span class="p">)</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span> <span class="c">#there is only one word </span>
			<span class="n">gene_array</span><span class="p">[</span><span class="n">gene_num</span><span class="p">]</span> <span class="o">=</span> <span class="n">split_words</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;*&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">g</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
			<span class="k">for</span> <span class="n">subv</span> <span class="ow">in</span> <span class="n">split_words</span><span class="p">:</span>
				<span class="c">#if we find a word with gene characters in it, it must be our gene </span>
				<span class="k">if</span> <span class="s">&#39;*&#39;</span> <span class="ow">in</span> <span class="n">subv</span> <span class="ow">or</span> <span class="s">&#39;-&#39;</span> <span class="ow">in</span> <span class="n">subv</span><span class="p">:</span>						
					<span class="c">#extract everything before &#39;*&#39;</span>
					<span class="n">g</span> <span class="o">=</span> <span class="n">subv</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;*&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
					<span class="k">break</span>
			<span class="n">gene_array</span><span class="p">[</span><span class="n">gene_num</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span>	
	<span class="k">return</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_array</span><span class="p">)</span>
	</div>
<div class="viewcode-block" id="get_top_genes"><a class="viewcode-back" href="../VHVLPairing.html#immunogrep_singlechain_cdr3_cluster.get_top_genes">[docs]</a><span class="k">def</span> <span class="nf">get_top_genes</span><span class="p">(</span><span class="n">gene_string</span><span class="p">,</span><span class="n">gene_scores</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>

<span class="sd">		Return a list of &#39;top&#39; unique genes</span>
<span class="sd">		Go through the reported list of genes and only report genes whose scores are equal to the top score</span>
<span class="sd">		</span>
<span class="sd">		Assumptions:	</span>
<span class="sd">			We err on the side of consider &#39;more genes&#39; than less</span>
<span class="sd">			1) If there is no score provided for a field, then consider all genes equally </span>
<span class="sd">			2) If only a single gene score is provided, then again, assume all reported genes have equal alignment (i.e. IMGT reporting)</span>
<span class="sd">			3) For all other cases where there is more than one gene score provided, then traverse the scores and only report genes whose score = top score</span>
<span class="sd">	</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="k">if</span> <span class="ow">not</span> <span class="n">gene_string</span><span class="p">:</span>
		<span class="k">return</span> <span class="p">[]</span>
	<span class="c">#gene_string should only be genes not alleles. this should have been handled in the add_to_dict_memory_safe function </span>
	<span class="n">gene_array</span> <span class="o">=</span> <span class="n">gene_string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)</span>
	
	<span class="c">#no scores provided, so we assume they are all equal </span>
	<span class="k">if</span> <span class="ow">not</span> <span class="n">gene_scores</span><span class="p">:</span>
		<span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">gene_array</span><span class="p">))</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">try</span><span class="p">:</span>			
			<span class="n">gene_numbers</span> <span class="o">=</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_scores</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)]</span>
		<span class="k">except</span><span class="p">:</span>
			<span class="c">#gene scores are not numbers</span>
			<span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">gene_array</span><span class="p">))</span>

		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_numbers</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
			<span class="c">#IMGT will report only one gene score for multiple genes. we assume these genes are idetnically scored then </span>
			<span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">gene_array</span><span class="p">))</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">top_score</span> <span class="o">=</span> <span class="n">gene_numbers</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
			<span class="n">top_gene_list</span> <span class="o">=</span> <span class="p">[]</span>
			<span class="k">for</span> <span class="n">n</span><span class="p">,</span><span class="n">s</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_numbers</span><span class="p">):</span>
				<span class="k">if</span> <span class="n">s</span><span class="o">&lt;</span><span class="n">top_score</span><span class="p">:</span>
					<span class="k">break</span>
				<span class="n">top_gene_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_array</span><span class="p">[</span><span class="n">n</span><span class="p">])</span>
			<span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">top_gene_list</span><span class="p">))</span>
						</div>
<div class="viewcode-block" id="parse_sorted_paired_file"><a class="viewcode-back" href="../VHVLPairing.html#immunogrep_singlechain_cdr3_cluster.parse_sorted_paired_file">[docs]</a><span class="k">def</span> <span class="nf">parse_sorted_paired_file</span><span class="p">(</span><span class="n">pairing_temp_file</span><span class="p">,</span><span class="n">suffix</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	</span>
<span class="sd">		Function for collapsing by CDR3 sequence</span>
<span class="sd">		We assume that the input file (pairing_temp_file) is a tab delimited file whose columns match the variable ab_field_loc		</span>
<span class="sd">		</span>
<span class="sd">	</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">ab_field_loc</span> <span class="o">=</span> <span class="n">get_ab_field_loc</span><span class="p">()</span>
	<span class="k">global</span> <span class="n">codeDict_output_dict</span>
	<span class="k">global</span> <span class="n">dict_summary</span>
	<span class="k">global</span> <span class="n">collapsed_output_file</span>
	<span class="k">global</span> <span class="n">usearch_cluster_file</span>
	
	<span class="c">#parent_folder = &#39;/&#39;.join(dict_summary.split(&#39;/&#39;)[:-1])+&#39;/&#39;</span>
	<span class="c">#temp_barcode_file = parent_folder+&#39;cdrh3_l3_barcodes.txt&#39;</span>
	
	<span class="n">my_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">pairing_temp_file</span><span class="p">)</span><span class="c"># &#39;/&#39;.join(pairing_temp_file.split(&#39;/&#39;)[:-1])+&#39;/&#39;</span>

	<span class="k">print</span><span class="p">(</span><span class="s">&#39;Will now sort file by CDR3 sequence&#39;</span><span class="p">)</span>
	<span class="c">#fullcode_column = num_elem_stored+1#last column number in file should be equal to the number of elements stored (in ab_field_loc) +1</span>
	<span class="c">#cdr3_column = ab_field_loc[&#39;CDR3&#39;]+1</span>
	<span class="c">#sort all sequences sent to file by the full code field (the last column in file) 		</span>
	<span class="c">#subprocess.call(&#39;&#39;&#39;sort -T &quot;{2}&quot; -t &#39;\t&#39; &quot;{0}&quot; -k{1}&gt;&quot;{0}.sorted&quot; &#39;&#39;&#39;.format(pairing_temp_file,str(cdr3_column),my_folder), shell=True)		</span>
	<span class="c">#os.remove(pairing_temp_file)</span>
	<span class="c">#os.rename(pairing_temp_file+&#39;.sorted&#39;,pairing_temp_file)</span>
		
	<span class="n">mapping_dict</span> <span class="o">=</span> <span class="p">{}</span>	
	<span class="n">r_count</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="n">index_loc</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;VDJ&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span><span class="s">&#39;VJ&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">}</span>	

	<span class="n">dict_summary</span><span class="o">+=</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="n">suffix</span><span class="o">+</span><span class="s">&#39;.txt&#39;</span>
	<span class="n">collapsed_output_file</span><span class="o">+=</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="n">suffix</span><span class="o">+</span><span class="s">&#39;.txt&#39;</span>
	<span class="n">usearch_cluster_file</span><span class="o">+=</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="n">suffix</span><span class="o">+</span><span class="s">&#39;.txt&#39;</span>
	<span class="c">#codeDict_save=open(codeDict_output_dict,&#39;w&#39;)</span>
	<span class="n">write_dict_summary</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">dict_summary</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>
	<span class="n">write_dict_summary_unsorted</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">dict_summary</span><span class="o">+</span><span class="s">&#39;.unsorted.txt&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>
	<span class="c">#collapsed_cdr3_barcode_file = open(temp_barcode_file,&#39;w&#39;)	</span>
	<span class="n">cdrh3_l3_barcode_dict</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
	<span class="c">#codeDict_save.write(&#39;{&#39;)#we will save the variable codeDict as a JSON var to file. JSON files start with { and end with }. &#39;,&#39; will be used to save each R1-R2 pair</span>
	
	<span class="c">#read through current filtered file and store counts of unique CDRH3. We dont</span>
	<span class="c">#also create the &#39;line-by-line&#39; UNSORTED and line-by-line SORTED files </span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">pairing_temp_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">unsorted_file</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">unsorted_file</span><span class="p">:</span>
			<span class="n">line</span><span class="o">=</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\r\n</span><span class="s">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
			<span class="n">fullcode</span><span class="o">=</span><span class="n">line</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>			
			<span class="k">if</span> <span class="ow">not</span> <span class="n">fullcode</span><span class="p">:</span>
				<span class="c">#shouldnt happen, but check anyway...</span>
				<span class="k">continue</span>				
			<span class="n">temp_array</span> <span class="o">=</span> <span class="n">line</span>			
			<span class="n">rtype</span> <span class="o">=</span> <span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;CHAIN_CALL&#39;</span><span class="p">]]</span>			
			<span class="n">pos</span> <span class="o">=</span> <span class="n">index_loc</span><span class="p">[</span><span class="n">rtype</span><span class="p">]</span>
			<span class="n">array</span> <span class="o">=</span> <span class="p">[[],[],</span><span class="s">&#39;&#39;</span><span class="p">]</span>
			<span class="n">array</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">field</span><span class="p">:</span><span class="n">temp_array</span><span class="p">[</span><span class="n">index_val</span><span class="p">]</span> <span class="k">for</span> <span class="n">field</span><span class="p">,</span><span class="n">index_val</span> <span class="ow">in</span> <span class="n">ab_field_loc</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()}</span>
			<span class="n">barcode_cdr3</span> <span class="o">=</span><span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;CDR3_SEQ&#39;</span><span class="p">]]</span>
			<span class="n">write_line</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">array</span><span class="p">)</span>
			<span class="n">write_dict_summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">barcode_cdr3</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">+</span><span class="n">write_line</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>						 
			<span class="n">write_dict_summary_unsorted</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">write_line</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>						
			<span class="n">cdrh3_l3_barcode_dict</span><span class="p">[</span><span class="n">barcode_cdr3</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>							
												
	<span class="n">write_dict_summary</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
	<span class="n">write_dict_summary_unsorted</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
	
	<span class="c">#Now sort the write_dict_summary file by CDRH3-L3 barcode sequences</span>
	<span class="k">print</span><span class="p">(</span><span class="s">&#39;Sorting paired sequences by unique CDR3 sequences&#39;</span><span class="p">)</span>
	<span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="s">&#39;&#39;&#39;sort -T &quot;{1}&quot; -t &#39;</span><span class="se">\t</span><span class="s">&#39; &quot;{0}&quot; -k1,1 &gt; &quot;{0}.sorted&quot; &#39;&#39;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">dict_summary</span><span class="p">),</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">my_folder</span><span class="p">)),</span> <span class="n">shell</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
	<span class="c">#Now open the file again and keep track of unique CDRH3-CDRL3 sequence data. Keep and updated list of SHM to calculate the averate and standard deviation. </span>
	<span class="c">#Also keep track of which line has the MOST data. i.e. has all vgenes and isotype information. By default we will store this information as the &#39;unique&#39; sequence. </span>
	<span class="k">print</span><span class="p">(</span><span class="s">&#39;Collapsing sequences by unique CDR3. Finding average SHM and most occurring &quot;Gene Mode&quot; for unique sequences&#39;</span><span class="p">)</span>
	<span class="n">group_barcode</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
	<span class="n">barcode_count</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="n">summed_counts</span> <span class="o">=</span> <span class="mi">0</span> 
	<span class="n">collapsed_output</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">collapsed_output_file</span><span class="o">+</span><span class="s">&#39;.presorted&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>
	
	<span class="n">old_summary</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dict_summary</span><span class="o">+</span><span class="s">&#39;.sorted&#39;</span><span class="p">)</span>
	<span class="n">new_dict_summary</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">dict_summary</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>

	<span class="n">l</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="k">while</span> <span class="bp">True</span><span class="p">:</span>
		<span class="k">try</span><span class="p">:</span>
			<span class="n">each_seq_line</span> <span class="o">=</span> <span class="n">old_summary</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
		<span class="k">except</span><span class="p">:</span>
			<span class="k">break</span>
		<span class="k">if</span> <span class="ow">not</span> <span class="n">each_seq_line</span><span class="p">:</span>
			<span class="k">break</span>
		<span class="n">l</span><span class="o">+=</span><span class="mi">1</span>
		<span class="n">each_seq_line</span><span class="o">=</span><span class="n">each_seq_line</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\r\n</span><span class="s">&#39;</span><span class="p">)</span>
		<span class="n">each_seq_line</span> <span class="o">=</span> <span class="n">each_seq_line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
		<span class="n">barcode_cdr_h3_l3</span> <span class="o">=</span> <span class="n">each_seq_line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
		<span class="n">group_barcode</span> <span class="o">=</span> <span class="n">barcode_cdr_h3_l3</span>
		<span class="n">num_seqs</span> <span class="o">=</span> <span class="n">cdrh3_l3_barcode_dict</span><span class="p">[</span><span class="n">barcode_cdr_h3_l3</span><span class="p">]</span>
		<span class="n">current_seq</span> <span class="o">=</span> <span class="mi">1</span>
		<span class="n">remaining_data</span> <span class="o">=</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">each_seq_line</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
		<span class="n">new_dict_summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num_seqs</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">+</span><span class="n">remaining_data</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>							
		
		<span class="n">seed_barcode_data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">remaining_data</span><span class="p">)</span>
		<span class="n">vdj_data</span> <span class="o">=</span> <span class="n">seed_barcode_data</span><span class="p">[</span><span class="n">index_loc</span><span class="p">[</span><span class="s">&#39;VDJ&#39;</span><span class="p">]]</span>
		<span class="n">vj_data</span> <span class="o">=</span> <span class="n">seed_barcode_data</span><span class="p">[</span><span class="n">index_loc</span><span class="p">[</span><span class="s">&#39;VJ&#39;</span><span class="p">]]</span>
		
		<span class="k">if</span> <span class="ow">not</span> <span class="n">vdj_data</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">vj_data</span><span class="p">:</span>
			<span class="k">continue</span>
		<span class="k">elif</span> <span class="n">vdj_data</span><span class="p">:</span>
			<span class="n">data</span> <span class="o">=</span> <span class="n">vdj_data</span>
		<span class="k">elif</span> <span class="n">vj_data</span><span class="p">:</span>
			<span class="n">data</span> <span class="o">=</span> <span class="n">vj_data</span>
		<span class="n">vdj_data</span> <span class="o">=</span> <span class="bp">None</span>
		<span class="n">vj_data</span> <span class="o">=</span> <span class="bp">None</span>

		<span class="n">seed_barcode_data</span> <span class="o">=</span> <span class="n">data</span>

		<span class="n">num_not_empty</span> <span class="o">=</span> <span class="mi">0</span>
		<span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">t</span> <span class="ow">and</span> <span class="n">t</span><span class="o">!=</span><span class="s">&quot;none&quot;</span> <span class="ow">and</span> <span class="n">t</span><span class="o">!=</span><span class="s">&quot;N/A&quot;</span><span class="p">:</span>
				<span class="n">num_not_empty</span><span class="o">+=</span><span class="mi">1</span>
						
		<span class="n">most_fields</span> <span class="o">=</span> <span class="n">num_not_empty</span>
		<span class="n">shm_data</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span>
		
		<span class="n">mut_found</span><span class="o">=</span><span class="bp">False</span>
		
		<span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="s">&#39;MUT&#39;</span><span class="p">]</span><span class="o">!=</span><span class="s">&#39;none&#39;</span><span class="p">:</span>
			<span class="n">shm_val</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;MUT&#39;</span><span class="p">])</span>
			<span class="n">mut_found</span><span class="o">=</span><span class="bp">True</span>
			<span class="n">shm_data</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="n">shm_val</span><span class="p">,</span><span class="nb">pow</span><span class="p">(</span><span class="n">shm_val</span><span class="p">,</span><span class="mi">2</span><span class="p">)]</span>		
		
		<span class="n">vgene_counts</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
		<span class="n">jgene_counts</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
		<span class="n">dgene_counts</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
								
		<span class="c">#for all of the V(D)J genes in the pair, determine which genes are the top gene hits. return these genes as a list and update the current gene counts to the variable </span>
		<span class="c">#at the end of the group, we will determine the mode/most likely gene. for now, we only chose the first instance of a top count </span>
		<span class="k">for</span> <span class="n">top_genes</span> <span class="ow">in</span> <span class="n">get_top_genes</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;VGENE&#39;</span><span class="p">],</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;VSCORES&#39;</span><span class="p">]):</span>			
			<span class="n">vgene_counts</span><span class="p">[</span><span class="n">top_genes</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>
		<span class="k">for</span> <span class="n">top_genes</span> <span class="ow">in</span> <span class="n">get_top_genes</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;DGENE&#39;</span><span class="p">],</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;DSCORES&#39;</span><span class="p">]):</span>		
			<span class="n">dgene_counts</span><span class="p">[</span><span class="n">top_genes</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>
		<span class="k">for</span> <span class="n">top_genes</span> <span class="ow">in</span> <span class="n">get_top_genes</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;JGENE&#39;</span><span class="p">],</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;JSCORES&#39;</span><span class="p">]):</span>
			<span class="n">jgene_counts</span><span class="p">[</span><span class="n">top_genes</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>
				
		<span class="c">#go through all remaining sequences in this file that contains the same CDRH3-CDRL3 PAIR and take the sum of the SHM and SHM^2 												</span>
		<span class="k">while</span> <span class="n">current_seq</span><span class="o">!=</span><span class="n">num_seqs</span><span class="p">:</span>		
			<span class="n">current_seq</span><span class="o">+=</span><span class="mi">1</span>			
			<span class="n">each_seq_line</span> <span class="o">=</span> <span class="n">old_summary</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
			<span class="n">each_seq_line</span><span class="o">=</span><span class="n">each_seq_line</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\r\n</span><span class="s">&#39;</span><span class="p">)</span>
			<span class="n">each_seq_line</span> <span class="o">=</span> <span class="n">each_seq_line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
			<span class="n">barcode_cdr_h3_l3</span> <span class="o">=</span> <span class="n">each_seq_line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
			<span class="k">if</span> <span class="n">barcode_cdr_h3_l3</span><span class="o">!=</span><span class="n">group_barcode</span><span class="p">:</span>
				<span class="k">print</span> <span class="s">&#39;error matching barocdes&#39;</span>
				<span class="k">print</span> <span class="n">group_barcode</span>
				<span class="k">print</span> <span class="n">barcode_cdr_h3_l3</span>
				<span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;problem collapsing by unique cdrh3-cdrl3&#39;</span><span class="p">)</span>																
			
			<span class="n">remaining_data</span> <span class="o">=</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">each_seq_line</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
			<span class="n">new_dict_summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num_seqs</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">+</span><span class="n">remaining_data</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>							
			
			<span class="n">barcode_data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">remaining_data</span><span class="p">)</span>
			<span class="n">vdj_data</span> <span class="o">=</span> <span class="n">barcode_data</span><span class="p">[</span><span class="n">index_loc</span><span class="p">[</span><span class="s">&#39;VDJ&#39;</span><span class="p">]]</span>
			<span class="n">vj_data</span> <span class="o">=</span> <span class="n">barcode_data</span><span class="p">[</span><span class="n">index_loc</span><span class="p">[</span><span class="s">&#39;VJ&#39;</span><span class="p">]]</span>
			
		
			<span class="k">if</span> <span class="ow">not</span> <span class="n">vdj_data</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">vj_data</span><span class="p">:</span>
				<span class="k">continue</span>
			<span class="k">elif</span> <span class="n">vdj_data</span><span class="p">:</span>
				<span class="n">data</span> <span class="o">=</span> <span class="n">vdj_data</span>
			<span class="k">elif</span> <span class="n">vj_data</span><span class="p">:</span>
				<span class="n">data</span> <span class="o">=</span> <span class="n">vj_data</span>
			<span class="n">vdj_data</span> <span class="o">=</span> <span class="bp">None</span>
			<span class="n">vj_data</span> <span class="o">=</span> <span class="bp">None</span>


			<span class="n">num_not_empty</span> <span class="o">=</span> <span class="mi">0</span>
			<span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
				<span class="k">if</span> <span class="n">t</span> <span class="ow">and</span> <span class="n">t</span><span class="o">!=</span><span class="s">&quot;none&quot;</span> <span class="ow">and</span> <span class="n">t</span><span class="o">!=</span><span class="s">&quot;N/A&quot;</span><span class="p">:</span>
					<span class="n">num_not_empty</span><span class="o">+=</span><span class="mi">1</span>
			
			<span class="k">if</span> <span class="n">num_not_empty</span><span class="o">&gt;</span><span class="n">most_fields</span><span class="p">:</span>
				<span class="c">#if you find more non empty fields in this sequence, then it should be treated as the &#39;seed&#39;</span>
				<span class="n">seed_barcode_data</span> <span class="o">=</span> <span class="n">data</span>
				<span class="n">most_fields</span> <span class="o">=</span> <span class="n">num_not_empty</span>
			
			<span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="s">&#39;MUT&#39;</span><span class="p">]</span><span class="o">!=</span><span class="s">&#39;none&#39;</span><span class="p">:</span>
				<span class="n">shm_val</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;MUT&#39;</span><span class="p">])</span>
				<span class="n">mut_found</span><span class="o">=</span><span class="bp">True</span>					
				<span class="n">shm_data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>
				<span class="n">shm_data</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">+=</span><span class="n">shm_val</span>
				<span class="n">shm_data</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">+=</span><span class="nb">pow</span><span class="p">(</span><span class="n">shm_val</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
			
			<span class="c">#for all of the V(D)J genes in the pair, determine which genes are the top gene hits. return these genes as a list and update the current gene counts to the variable </span>
			<span class="c">#at the end of the group, we will determine the mode/most likely gene. for now, we only chose the first instance of a top count 			</span>
			<span class="k">for</span> <span class="n">top_genes</span> <span class="ow">in</span> <span class="n">get_top_genes</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;VGENE&#39;</span><span class="p">],</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;VSCORES&#39;</span><span class="p">]):</span>			
				<span class="n">vgene_counts</span><span class="p">[</span><span class="n">top_genes</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>
			<span class="k">for</span> <span class="n">top_genes</span> <span class="ow">in</span> <span class="n">get_top_genes</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;DGENE&#39;</span><span class="p">],</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;DSCORES&#39;</span><span class="p">]):</span>		
				<span class="n">dgene_counts</span><span class="p">[</span><span class="n">top_genes</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>
			<span class="k">for</span> <span class="n">top_genes</span> <span class="ow">in</span> <span class="n">get_top_genes</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;JGENE&#39;</span><span class="p">],</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;JSCORES&#39;</span><span class="p">]):</span>
				<span class="n">jgene_counts</span><span class="p">[</span><span class="n">top_genes</span><span class="p">]</span><span class="o">+=</span><span class="mi">1</span>
							
		<span class="k">if</span> <span class="n">mut_found</span><span class="p">:</span>
			<span class="n">mut_sum</span> <span class="o">=</span> <span class="n">shm_data</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
			<span class="n">mut_sum_sq</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">shm_data</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="mi">6</span><span class="p">)</span>
			<span class="n">mut_count</span> <span class="o">=</span> <span class="n">shm_data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>				
			<span class="n">sum_sq</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">pow</span><span class="p">(</span><span class="n">mut_sum</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span><span class="o">/</span><span class="n">mut_count</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>		
			<span class="c">#average shm</span>
			<span class="n">mut_avg</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">mut_sum</span><span class="o">/</span><span class="n">mut_count</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
			<span class="c">#shm variance -&gt; sum of squares formulat =&gt; sum(vals^2)-(sum(vals)^2/counts)</span>
			<span class="k">if</span> <span class="n">mut_count</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
				<span class="c">#mut_var = round(pow((mut_sum_sq-(pow(mut_sum,2)/mut_count))/(mut_count-1),0.5),3)</span>
				<span class="n">mut_var</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">pow</span><span class="p">((</span><span class="n">mut_sum_sq</span><span class="o">-</span><span class="n">sum_sq</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">mut_count</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span><span class="mf">0.5</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">mut_var</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">mut_avg</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span>
			<span class="n">mut_var</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span>
			<span class="n">mut_sum</span> <span class="o">=</span> <span class="mi">0</span>
			<span class="n">mut_sum_sq</span> <span class="o">=</span> <span class="mi">0</span>
			<span class="n">mut_count</span> <span class="o">=</span> <span class="mi">0</span>					
		
		<span class="n">data</span> <span class="o">=</span> <span class="n">seed_barcode_data</span>
		<span class="n">cdrlen</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s">&#39;CDR3_LEN&#39;</span><span class="p">]</span>

						
		<span class="n">summed_counts</span><span class="o">+=</span><span class="n">num_seqs</span>
		
		<span class="c">##store the FIRST gene matching MAX MODE of the V(D)J genes </span>
		<span class="c">#vdj_data[&#39;VGENE&#39;] = max(vgene_counts_vdj, key=vgene_counts_vdj.get)</span>
		<span class="c">#vdj_data[&#39;DGENE&#39;] = max(dgene_counts_vdj, key=dgene_counts_vdj.get)</span>
		<span class="c">#vdj_data[&#39;JGENE&#39;] = max(jgene_counts_vdj, key=jgene_counts_vdj.get)</span>
		<span class="c">#vj_data[&#39;VGENE&#39;] = max(vgene_counts_vj, key=vgene_counts_vj.get)		</span>
		<span class="c">#vj_data[&#39;JGENE&#39;] = max(jgene_counts_vj, key=jgene_counts_vj.get)</span>

		<span class="c">##store the ALL genes matching MAX MODE of the V(D)J genes </span>
		<span class="k">if</span> <span class="n">vgene_counts</span><span class="p">:</span>
			<span class="n">max_item</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">vgene_counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
			<span class="n">data</span><span class="p">[</span><span class="s">&#39;VGENE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">g</span> <span class="k">for</span> <span class="n">g</span><span class="p">,</span><span class="n">num</span> <span class="ow">in</span> <span class="n">vgene_counts</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()</span> <span class="k">if</span> <span class="n">num</span><span class="o">==</span><span class="n">max_item</span><span class="p">])</span> <span class="k">if</span> <span class="n">max_item</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s">&#39;&#39;</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">data</span><span class="p">[</span><span class="s">&#39;VGENE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
		<span class="k">if</span> <span class="n">dgene_counts</span><span class="p">:</span>
			<span class="n">max_item</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">dgene_counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
			<span class="n">data</span><span class="p">[</span><span class="s">&#39;DGENE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">g</span> <span class="k">for</span> <span class="n">g</span><span class="p">,</span><span class="n">num</span> <span class="ow">in</span> <span class="n">dgene_counts</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()</span> <span class="k">if</span> <span class="n">num</span><span class="o">==</span><span class="n">max_item</span><span class="p">])</span> <span class="k">if</span> <span class="n">max_item</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s">&#39;&#39;</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">data</span><span class="p">[</span><span class="s">&#39;DGENE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
	
		<span class="k">if</span> <span class="n">jgene_counts</span><span class="p">:</span>
			<span class="n">max_item</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">jgene_counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
			<span class="n">data</span><span class="p">[</span><span class="s">&#39;JGENE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">g</span> <span class="k">for</span> <span class="n">g</span><span class="p">,</span><span class="n">num</span> <span class="ow">in</span> <span class="n">jgene_counts</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()</span> <span class="k">if</span> <span class="n">num</span><span class="o">==</span><span class="n">max_item</span><span class="p">])</span> <span class="k">if</span> <span class="n">max_item</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s">&#39;&#39;</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">data</span><span class="p">[</span><span class="s">&#39;JGENE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
		
		<span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="s">&#39;CHAIN_CALL&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">==</span><span class="s">&#39;VDJ&#39;</span><span class="p">:</span>
			<span class="c">#the variable ab_field_loc stores the index position for each antibody region in the array </span>
			<span class="n">CDRH3</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;CDR3_SEQ&#39;</span><span class="p">]</span>
			<span class="n">VHgene</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;VGENE&#39;</span><span class="p">]</span>
			<span class="n">DHgene</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;DGENE&#39;</span><span class="p">]</span>
			<span class="n">JHgene</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;JGENE&#39;</span><span class="p">]</span>
			<span class="n">CDRL3</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">VLgene</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">JLgene</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">IgH</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;LOCUS&#39;</span><span class="p">]</span>
			<span class="n">IgL</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">h_iso</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s">&#39;ISOTYPE&#39;</span><span class="p">]</span>
			<span class="n">l_iso</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
			<span class="n">Hlen</span><span class="o">=</span><span class="n">cdrlen</span>
			<span class="n">Llen</span><span class="o">=</span><span class="s">&#39;&#39;</span>		
			<span class="k">try</span><span class="p">:</span> 
				<span class="n">CDRH3_trans</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">CDRH3</span><span class="p">,</span><span class="n">generic_dna</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
			<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> 
				<span class="n">CDRH3</span><span class="o">=</span> <span class="s">&#39;translation error&#39;</span>
				<span class="k">print</span><span class="p">(</span><span class="s">&#39;Error translating: &#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
			<span class="n">CDRL3_trans</span><span class="o">=</span><span class="s">&#39;&#39;</span>

			<span class="n">Hmut_avg</span> <span class="o">=</span> <span class="n">mut_avg</span>
			<span class="n">Hmut_var</span> <span class="o">=</span> <span class="n">mut_var</span>
			<span class="n">Lmut_avg</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
			<span class="n">Lmut_var</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
			<span class="n">Hmut_sum</span><span class="o">=</span><span class="n">mut_sum</span>
			<span class="n">Hmut_sum_sq</span><span class="o">=</span><span class="n">mut_sum_sq</span>
			<span class="n">Hmut_count</span><span class="o">=</span><span class="n">mut_count</span>
			<span class="n">Lmut_sum</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">Lmut_sum_sq</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">Lmut_count</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>

		<span class="k">elif</span> <span class="n">data</span><span class="p">[</span><span class="s">&#39;CHAIN_CALL&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">==</span><span class="s">&#39;VJ&#39;</span><span class="p">:</span>
			<span class="c">#the variable ab_field_loc stores the index position for each antibody region in the array </span>
			<span class="n">CDRH3</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">VHgene</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">DHgene</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">JHgene</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">CDRL3</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;CDR3_SEQ&#39;</span><span class="p">]</span>
			<span class="n">VLgene</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;VGENE&#39;</span><span class="p">]</span>
			<span class="n">JLgene</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;JGENE&#39;</span><span class="p">]</span>
			<span class="n">IgH</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">IgL</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;LOCUS&#39;</span><span class="p">]</span>
			<span class="n">h_iso</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
			<span class="n">l_iso</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s">&#39;LOCUS&#39;</span><span class="p">]</span>
			<span class="n">Hlen</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
			<span class="n">Llen</span> <span class="o">=</span> <span class="n">cdrlen</span>

			<span class="n">Hmut_avg</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
			<span class="n">Hmut_var</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
			<span class="n">Lmut_avg</span> <span class="o">=</span> <span class="n">mut_avg</span>
			<span class="n">Lmut_var</span> <span class="o">=</span> <span class="n">mut_var</span>
			<span class="n">Hmut_sum</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">Hmut_sum_sq</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">Hmut_count</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			<span class="n">Lmut_sum</span><span class="o">=</span><span class="n">mut_sum</span>
			<span class="n">Lmut_sum_sq</span><span class="o">=</span><span class="n">mut_sum</span>
			<span class="n">Lmut_count</span> <span class="o">=</span> <span class="n">mut_count</span>
						
			<span class="k">try</span><span class="p">:</span> 
				<span class="n">CDRL3_trans</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">CDRL3</span><span class="p">,</span><span class="n">generic_dna</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
			<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> 
				<span class="n">CDRL3</span><span class="o">=</span> <span class="s">&#39;translation error&#39;</span>
				<span class="k">print</span><span class="p">(</span><span class="s">&#39;Error translating: &#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
			<span class="n">CDRH3_trans</span><span class="o">=</span><span class="s">&#39;&#39;</span>
			

		<span class="n">results</span> <span class="o">=</span> <span class="p">[</span><span class="n">num_seqs</span><span class="p">,</span><span class="n">CDRH3_trans</span><span class="p">,</span><span class="n">CDRL3_trans</span><span class="p">,</span><span class="n">CDRH3</span><span class="p">,</span><span class="n">CDRL3</span><span class="p">,</span><span class="n">VHgene</span><span class="p">,</span><span class="n">DHgene</span><span class="p">,</span><span class="n">JHgene</span><span class="p">,</span><span class="n">VLgene</span><span class="p">,</span><span class="n">JLgene</span><span class="p">,</span><span class="n">IgH</span><span class="p">,</span><span class="n">IgL</span><span class="p">,</span><span class="n">Hmut_avg</span><span class="p">,</span><span class="n">Hmut_var</span><span class="p">,</span><span class="n">Lmut_avg</span><span class="p">,</span><span class="n">Lmut_var</span><span class="p">,</span><span class="n">Hlen</span><span class="p">,</span><span class="n">Llen</span><span class="p">,</span><span class="n">h_iso</span><span class="p">,</span><span class="n">l_iso</span><span class="p">,</span><span class="n">Hmut_sum</span><span class="p">,</span><span class="n">Hmut_sum_sq</span><span class="p">,</span><span class="n">Hmut_count</span><span class="p">,</span><span class="n">Lmut_sum</span><span class="p">,</span><span class="n">Lmut_sum_sq</span><span class="p">,</span><span class="n">Lmut_count</span><span class="p">]</span>
		<span class="n">results</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">results</span><span class="p">]</span>
			
		<span class="n">collapsed_output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
			
	<span class="n">old_summary</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
	<span class="n">new_dict_summary</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>	
	
	<span class="n">collapsed_output</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
	
	<span class="k">print</span><span class="p">(</span><span class="s">&#39;Compiling collapsed reads file: sorting unique CDRH3-CDRL3 sequences by their respective counts and exporting to FASTA file&#39;</span><span class="p">)</span>
	
	<span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="s">&#39;&#39;&#39;sort -T &quot;{1}&quot; -t &#39;</span><span class="se">\t</span><span class="s">&#39; &quot;{0}.presorted&quot; -k1,1nr &gt;&quot;{0}&quot; &#39;&#39;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">collapsed_output_file</span><span class="p">,</span><span class="n">my_folder</span><span class="p">),</span> <span class="n">shell</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
	
	<span class="c">#convert output to fasta file using AWK, ONLY print counts ($1) above 1 </span>
	
	<span class="k">if</span> <span class="n">suffix</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">==</span><span class="s">&#39;VDJ&#39;</span><span class="p">:</span>
		<span class="n">fasta_seq</span> <span class="o">=</span> <span class="s">&quot;$4&quot;</span> <span class="c">#$4 =&gt; nucleotide CDRH3</span>
	<span class="k">elif</span> <span class="n">suffix</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">==</span><span class="s">&#39;VJ&#39;</span><span class="p">:</span>		
		<span class="n">fasta_seq</span> <span class="o">=</span> <span class="s">&quot;$5&quot;</span> <span class="c">#$4 =&gt; nucleotide CDRL3</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;Error providing suffix field to function: &#39;</span><span class="o">+</span><span class="n">suffix</span><span class="p">)</span>


	<span class="n">awk_command</span><span class="o">=</span><span class="s">&#39;gawk -v offile=&quot;&#39;</span><span class="o">+</span><span class="n">usearch_cluster_file</span>
	<span class="n">awk_command</span><span class="o">+=</span><span class="s">&#39;&#39;&#39;&quot; &#39;BEGIN{OFS=&quot;:&quot;;FS=&quot;</span><span class="se">\t</span><span class="s">&quot;};</span>
<span class="s">					(int($1)&gt;1&amp;&amp;&#39;&#39;&#39;</span><span class="o">+</span><span class="n">fasta_seq</span><span class="o">+</span><span class="s">&#39;&#39;&#39;!=&quot;&quot;){s=$1;for(i=2;i&lt;=NF;i++)s=s&quot;&quot;OFS&quot;&quot;$i; print&quot;&gt;&quot;s&gt;offile;print &#39;&#39;&#39;</span><span class="o">+</span><span class="n">fasta_seq</span><span class="o">+</span><span class="s">&#39;&#39;&#39;&gt;offile;}&#39; &#39;&#39;&#39;</span><span class="o">+</span><span class="n">collapsed_output_file</span>
	
	<span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">awk_command</span><span class="p">,</span><span class="n">shell</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
	
	<span class="c">#if there are no VH-VL pairs found, then the awk command above will not create an empty file.</span>
	<span class="c">#so to account for that we will do a hack and create an empty file </span>
	<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">usearch_cluster_file</span><span class="p">):</span>
		<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">usearch_cluster_file</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">w</span><span class="p">:</span>
			<span class="k">pass</span>
	
	<span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">dict_summary</span><span class="o">+</span><span class="s">&#39;.sorted&#39;</span><span class="p">)</span>
	<span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">collapsed_output_file</span><span class="o">+</span><span class="s">&#39;.presorted&#39;</span><span class="p">)</span>
	
	<span class="k">print</span><span class="p">(</span><span class="s">&#39;Creating a sorted line-by-line cluster file&#39;</span><span class="p">)</span>
	<span class="n">os</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">dict_summary</span><span class="p">,</span><span class="n">dict_summary</span><span class="o">+</span><span class="s">&#39;.presorted&#39;</span><span class="p">)</span>	
	<span class="n">final_sort</span> <span class="o">=</span> <span class="s">&#39;&#39;&#39;sort -T &quot;{1}&quot; -t &#39;</span><span class="se">\t</span><span class="s">&#39; &quot;{0}.presorted&quot; -k1,1nr | gawk &#39;BEGIN{{FS=&quot;</span><span class="se">\t</span><span class="s">&quot;}};{{print $2 &gt; &quot;{0}&quot;}}&#39; &#39;&#39;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dict_summary</span><span class="p">,</span><span class="n">my_folder</span><span class="p">)</span>	
	<span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">final_sort</span><span class="p">,</span><span class="n">shell</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>	
	<span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">dict_summary</span><span class="o">+</span><span class="s">&#39;.presorted&#39;</span><span class="p">)</span>
	
	<span class="n">num_unique_h_l</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cdrh3_l3_barcode_dict</span><span class="p">)</span>
	<span class="n">cdrh3_l3_barcode_dict</span> <span class="o">=</span> <span class="p">{}</span>
	<span class="k">del</span>  <span class="n">cdrh3_l3_barcode_dict</span>
	<span class="n">gc</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
	
	
	<span class="k">return</span> <span class="p">[</span><span class="n">num_unique_h_l</span><span class="p">]</span>
	

<span class="c">#THIS IS another version of add_to_dict. Its functionality should be identical except it does not store everything inside of memory </span>
<span class="c">#The previous ADD_TO_DICT FUNCTION is in the main function after this one !!!</span>
<span class="c">#THIS FUNCTION DIFFERS FROM ADD_TO_DICT BELOW IN THAT IT WILL TRY TO MINIMIZE THE AMOUNT OF MEMORY REQUIRED FOR RUNNING ANALYSIS (it uses about 1/3 to 1/4th of the memory required by  add_to_dict</span>
<span class="c">#initialitests show that speed is not affected using this method. That is, its not slower than the previous add_to_dict method </span>
<span class="c">#THIS FUNCTION WILL READ THROUGH THE LIST OF FILES PROVIDED</span>
<span class="c">#SO INSTEAD OF STORING DATA FROM READS IN A DICTIONARY, IT WILL INSTEAD STORE THE RESULTS IN a single tab delimited file </span>
<span class="c">#this file will be parsed by the function parse_sorted_paired_file using a mixture of awk and python code (see parse_sorted_paired_file) function </span></div>
<span class="k">def</span> <span class="nf">add_to_dict_memory_safe</span><span class="p">(</span><span class="n">list_of_files</span><span class="p">,</span><span class="n">list_of_filetypes</span><span class="p">,</span><span class="n">required_field_names</span><span class="p">,</span><span class="n">productivity_function_call</span><span class="p">,</span><span class="n">parent_dir</span><span class="p">,</span><span class="n">analysis_method</span><span class="p">,</span><span class="n">include_vl_analysis</span><span class="p">):</span>	
	<span class="c">##DEFINING SOME VARIABLES##</span>

	<span class="c">#get global files names for opening files to write/use in this function </span>
	<span class="k">global</span> <span class="n">dict_summary</span>
	<span class="k">global</span> <span class="n">usearch_cluster_file</span>
	
	<span class="c">#set global variables for field names we require to run analysis </span>
	<span class="k">global</span> <span class="n">vgene_field</span>
	<span class="k">global</span> <span class="n">jgene_field</span>
	<span class="k">global</span> <span class="n">dgene_field</span>
	<span class="k">global</span> <span class="n">vgene_scores</span>
	<span class="k">global</span> <span class="n">jgene_scores</span>
	<span class="k">global</span> <span class="n">dgene_scores</span>
	<span class="k">global</span> <span class="n">recomb_field</span>
	<span class="k">global</span> <span class="n">seq_field</span>
	<span class="k">global</span> <span class="n">cdr3_field</span>
	<span class="k">global</span> <span class="n">header_field</span>
	<span class="k">global</span> <span class="n">mut_field</span>
	<span class="k">global</span> <span class="n">functionality_field</span>
	<span class="k">global</span> <span class="n">full_aa_field</span>
	<span class="k">global</span> <span class="n">full_nt_field</span>
	
	<span class="n">required_field_names</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="n">required_field_names</span><span class="p">)</span> <span class="c">#convert this to default dict so that not all fields have to be explicitly defined</span>
	
	<span class="c">#keys to that refer to field names. these are the key values we use for reading files </span>
	<span class="n">vgene_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;vgene&#39;</span><span class="p">]</span>
	<span class="n">jgene_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;jgene&#39;</span><span class="p">]</span>
	<span class="n">dgene_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;dgene&#39;</span><span class="p">]</span>
	<span class="n">seq_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;raw_seq_nt&#39;</span><span class="p">]</span>
	<span class="n">cdr3_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;cdr3_nt&#39;</span><span class="p">]</span>
	<span class="n">isotype_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;isotype&#39;</span><span class="p">]</span>
	<span class="n">header_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;seq_header&#39;</span><span class="p">]</span>
	<span class="n">mut_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;shm&#39;</span><span class="p">]</span>
	<span class="n">functionality_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;functionality&#39;</span><span class="p">]</span>
	<span class="n">full_aa_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;full_len_ab_aa&#39;</span><span class="p">]</span>
	<span class="n">full_nt_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;full_len_ab_nt&#39;</span><span class="p">]</span>
	<span class="n">vgene_scores</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;vgene_scores&#39;</span><span class="p">]</span>
	<span class="n">jgene_scores</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;jgene_scores&#39;</span><span class="p">]</span> 
	<span class="n">dgene_scores</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;dgene_scores&#39;</span><span class="p">]</span>
	<span class="n">recomb_field</span> <span class="o">=</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;recomb&#39;</span><span class="p">]</span>
	
	<span class="n">num_vdj</span> <span class="o">=</span> <span class="mi">0</span> 
	<span class="n">num_vj</span> <span class="o">=</span> <span class="mi">0</span> 

	<span class="c">#seq_field = required_field_names[&#39;sequence&#39;]</span>
	
	<span class="k">if</span> <span class="n">productivity_function_call</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span>
		<span class="n">productivity_function_call</span> <span class="o">=</span> <span class="n">CDR3Productivity</span>	<span class="c">#GeneralProductivity</span>
										
	<span class="c">#write_dict_summary = open(dict_summary,&#39;w&#39;)</span>
	<span class="c">#collapsed_save=open(collapsed_output_dict,&#39;w&#39;)	</span>
	<span class="c">#collapsed_output=open(collapsed_output_file,&#39;wb&#39;)</span>
	<span class="c">#error=open(error_file,&#39;wb&#39;)	</span>
	
	<span class="n">pairing_temp_file</span> <span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">dict_summary</span><span class="p">),</span><span class="s">&#39;temp_annotation_files&#39;</span><span class="p">)</span> <span class="c">#&#39;/&#39;.join(dict_summary.split(&#39;/&#39;)[:-1])+&#39;/temp_annotation_files.txt&#39;</span>
	<span class="n">temp_output_files_vdj</span>  <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">pairing_temp_file</span><span class="o">+</span><span class="s">&#39;.vdj.txt&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>		
	<span class="n">temp_output_files_vj</span>  <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">pairing_temp_file</span><span class="o">+</span><span class="s">&#39;.vj.txt&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">include_vl_analysis</span> <span class="k">else</span> <span class="bp">None</span>
	
	<span class="c">#temp_seq_data = open(pairing_temp_file,&#39;w&#39;)	</span>
		
	<span class="n">counter</span><span class="o">=</span><span class="mi">0</span>
		
	<span class="n">codeDictSeqs</span> <span class="o">=</span> <span class="p">{}</span> <span class="c">#stores whether or not a sequence was successfully found (i.e. passes all filters) =&gt; 0 = not successful, 1 = successful </span>
	<span class="n">collapsed_dict</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="c">#store counts for unique cdrH3/CDRL3 counts </span>
	
	<span class="n">nonpaired_len</span> <span class="o">=</span> <span class="mi">0</span>	
	<span class="n">no_cdr3_error</span> <span class="o">=</span> <span class="mi">0</span> <span class="c">#keep track of sequences which lack cdr3</span>
	<span class="n">invalid_chain</span> <span class="o">=</span> <span class="mi">0</span> <span class="c">#keep track of sequences which do not have a valid chain call based on the chain_call variable above </span>
	<span class="n">not_productive</span> <span class="o">=</span> <span class="mi">0</span> <span class="c">#keep track of sequences which are determined to be &#39;unproductive&#39;</span>
	<span class="n">total_seqs</span> <span class="o">=</span> <span class="mi">0</span> <span class="c">#keep track of all sequences</span>
	<span class="n">no_result</span> <span class="o">=</span> <span class="mi">0</span> <span class="c">#keep track of sequences that have no antibody information 	</span>
	<span class="n">passed_filter</span> <span class="o">=</span> <span class="mi">0</span> <span class="c">#keep track of sequences that pass the filters described above (no cdr3, unproductive, etc)	</span>
	
	<span class="n">ab_field_loc</span> <span class="o">=</span> <span class="n">get_ab_field_loc</span><span class="p">()</span>
	
	<span class="c">###VARIABLES DEFINED, CONTINUING WITH FUNCTION###</span>
	
	<span class="k">print</span> <span class="s">&#39;Reading all files at once and collapsing identical CDRH3-CDRL3 pairs into collapsed dict&#39;</span> 
	
	<span class="n">at_least_one_file_open</span> <span class="o">=</span> <span class="bp">True</span>
	
	<span class="c">#we will generate annotation files as we read through the files for pairing </span>
	<span class="n">list_of_file_reading</span> <span class="o">=</span> <span class="n">initialize_input_files</span><span class="p">(</span><span class="n">analysis_method</span><span class="p">,</span><span class="n">list_of_filetypes</span><span class="p">,</span><span class="n">list_of_files</span><span class="p">)</span>
	<span class="c"># we will read each of the files simultaneously (open all the files at once and read line by line)</span>
	<span class="k">while</span> <span class="n">at_least_one_file_open</span><span class="p">:</span>		
	<span class="c">#while counter&lt;200000:</span>
		<span class="n">counter</span><span class="o">+=</span><span class="mi">1</span>
		<span class="k">if</span> <span class="n">total_seqs</span><span class="o">%</span><span class="mi">100000</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
			<span class="k">print</span> <span class="s">&#39;Processed &#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">total_seqs</span><span class="p">)</span><span class="o">+</span><span class="s">&#39; sequences&#39;</span><span class="c">#: &#39;+str(counter)			</span>
		
		<span class="c">#keep track of how many files have been completely read</span>
		<span class="n">num_eof</span> <span class="o">=</span> <span class="mi">0</span> 
		
		<span class="c">#read each file line by line</span>
		<span class="k">for</span> <span class="n">fnum</span><span class="p">,</span><span class="n">reader</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">list_of_file_reading</span><span class="p">):</span> 
			<span class="k">if</span> <span class="n">reader</span><span class="o">.</span><span class="n">IFclass</span><span class="o">.</span><span class="n">eof</span><span class="p">:</span>
				<span class="n">num_eof</span><span class="o">+=</span><span class="mi">1</span>
				<span class="k">continue</span>						
				
			<span class="n">my_line</span> <span class="o">=</span> <span class="n">reader</span><span class="o">.</span><span class="n">IFclass</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
							
			<span class="k">if</span> <span class="ow">not</span> <span class="n">my_line</span><span class="p">:</span><span class="c">#probably end of file or just empty line			</span>
				<span class="k">continue</span>
			
			<span class="n">my_line</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="n">my_line</span><span class="p">)</span>
			<span class="n">h</span> <span class="o">=</span> <span class="n">my_line</span><span class="p">[</span><span class="n">header_field</span><span class="p">]</span>				
			<span class="k">if</span> <span class="ow">not</span> <span class="n">h</span><span class="p">:</span>								
				<span class="k">print</span> <span class="s">&#39;Error type 0: no sequence header&#39;</span>
				<span class="k">continue</span>
			
			<span class="p">[</span><span class="n">header</span><span class="p">,</span><span class="nb">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">GetHeaderInfo</span><span class="p">(</span><span class="n">my_line</span><span class="p">,</span><span class="n">header_field</span><span class="p">)</span>						
			<span class="n">seq</span> <span class="o">=</span> <span class="n">my_line</span><span class="p">[</span><span class="n">seq_field</span><span class="p">]</span>
			<span class="c">#get miseq code/read info </span>
			<span class="n">fullcode</span> <span class="o">=</span> <span class="s">&#39;:&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">header</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39; &#39;</span><span class="p">,</span><span class="s">&#39;_&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">)[</span><span class="mi">3</span><span class="p">:</span><span class="mi">7</span><span class="p">])</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>															
						
			<span class="n">total_seqs</span><span class="o">+=</span><span class="mi">1</span>												
			
			<span class="c">#no amino acid sequence found </span>
			<span class="k">if</span> <span class="ow">not</span> <span class="n">my_line</span><span class="p">[</span><span class="n">full_aa_field</span><span class="p">]:</span>			
				
				<span class="c">#we no longer require the variable full_aa_field to be &#39;non empty&#39;</span>
				<span class="k">if</span> <span class="ow">not</span> <span class="n">my_line</span><span class="p">[</span><span class="n">full_nt_field</span><span class="p">]:</span>					
					<span class="c">#no nucleotide sequence found either </span>
					<span class="c">#do not save this sequences results to file </span>
					<span class="c">#write to temperoary file annotation file, store that there is no full length sequence for this sequence, so no need to save its </span>
					<span class="c">#paired annotation information </span>
					<span class="c">#annotation_file_writing[fnum][&#39;buffer&#39;].write(&#39;\t&#39;.join([header,seq,id,&#39;&#39;,&#39;&#39;,&#39;&#39;])+&#39;\n&#39;)						</span>
					<span class="c">#no_result+=1</span>
					<span class="c">#continue</span>
					<span class="k">pass</span>
				<span class="k">else</span><span class="p">:</span>
					<span class="c">#aminoa acid sequence was not proviided, but a nucleotide sequence was provided </span>
					<span class="c">#translate nucloetid to amino acid..this can result in some potential problems with PRODUCTIVITY determination IF using GENERAL PRODUCTIVTY function rule </span>
					<span class="k">try</span><span class="p">:</span>
						<span class="n">end</span> <span class="o">=</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">my_line</span><span class="p">[</span><span class="n">full_nt_field</span><span class="p">])</span><span class="o">/</span><span class="mi">3</span><span class="p">)</span><span class="o">*</span><span class="mi">3</span>
						<span class="n">my_line</span><span class="p">[</span><span class="n">full_aa_field</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">my_line</span><span class="p">[</span><span class="n">full_nt_field</span><span class="p">][:</span><span class="n">end</span><span class="p">],</span><span class="n">generic_dna</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
					<span class="k">except</span><span class="p">:</span>
						<span class="k">print</span><span class="p">(</span><span class="s">&#39;Error type 2: problem translating amino acid =&gt; &#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">my_line</span><span class="p">[</span><span class="n">full_nt_field</span><span class="p">]))</span>
						<span class="c">#annotation_file_writing[fnum][&#39;buffer&#39;].write(&#39;\t&#39;.join([header,seq,id,&#39;&#39;,&#39;&#39;,&#39;&#39;])+&#39;\n&#39;)						</span>
						<span class="c">#no_result+=1</span>
						<span class="c">#continue</span>
						<span class="n">my_line</span><span class="p">[</span><span class="n">full_aa_field</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
						<span class="c">#no cdr3 found </span>
			<span class="k">if</span> <span class="ow">not</span><span class="p">(</span><span class="n">my_line</span><span class="p">[</span><span class="n">cdr3_field</span><span class="p">]):</span>											
				<span class="c">#write to temperoary file 				</span>
				<span class="n">no_cdr3_error</span><span class="o">+=</span><span class="mi">1</span>
				<span class="k">continue</span>				
									
			<span class="c">#maintain all vgenes as a list </span>
			<span class="c">#Process genes and only extract the GENE from each of the fields (that is ignore the alleles &#39;*01&#39;)</span>
			<span class="n">Vgene</span> <span class="o">=</span> <span class="n">ProcessGene</span><span class="p">(</span><span class="n">my_line</span><span class="p">[</span><span class="n">vgene_field</span><span class="p">])</span><span class="c">#.split(&#39;,&#39;)#[0]</span>
			<span class="n">Jgene</span> <span class="o">=</span> <span class="n">ProcessGene</span><span class="p">(</span><span class="n">my_line</span><span class="p">[</span><span class="n">jgene_field</span><span class="p">])</span><span class="c">#.split(&#39;,&#39;)#[0]			</span>
			<span class="n">Dgene</span> <span class="o">=</span> <span class="n">ProcessGene</span><span class="p">(</span><span class="n">my_line</span><span class="p">[</span><span class="n">dgene_field</span><span class="p">])</span><span class="c">#.split(&#39;,&#39;)#[0]				</span>
			<span class="c">#Unlike the above change for genes, only keep the first isotype in the list </span>
			<span class="n">isotype</span> <span class="o">=</span> <span class="n">my_line</span><span class="p">[</span><span class="n">isotype_field</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>			
			<span class="c">#figure out locus using vgene call 														</span>
		
			<span class="k">if</span> <span class="n">Vgene</span><span class="p">:</span>
				<span class="c">#use vgene to determine locus </span>
				<span class="n">locus</span> <span class="o">=</span> <span class="n">Vgene</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>						
			<span class="k">elif</span> <span class="n">Jgene</span><span class="p">:</span>
				<span class="c">#if no vgene is present, then use jgene</span>
				<span class="n">locus</span> <span class="o">=</span> <span class="n">Jgene</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>			
			<span class="k">else</span><span class="p">:</span>
				<span class="c">#ignore using D gene</span>
				<span class="n">locus</span><span class="o">=</span><span class="s">&#39;&#39;</span>			
			
			<span class="n">chain</span> <span class="o">=</span> <span class="s">&#39;N/A&#39;</span>			
		
			<span class="c">#the provided file defines recombination/VDJ or VJ chain </span>
			<span class="k">if</span> <span class="n">my_line</span><span class="p">[</span><span class="n">recomb_field</span><span class="p">]</span> <span class="ow">and</span> <span class="n">my_line</span><span class="p">[</span><span class="n">recomb_field</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s">&#39;VDJ&#39;</span><span class="p">,</span><span class="s">&#39;VJ&#39;</span><span class="p">]:</span>
				<span class="n">chain</span> <span class="o">=</span> <span class="n">my_line</span><span class="p">[</span><span class="n">recomb_field</span><span class="p">]</span>
			<span class="k">elif</span> <span class="n">locus</span><span class="p">:</span>				
				<span class="c">#shoudl return &#39;VDJ&#39; or &#39;VJ&#39; based on LOCUS var</span>
				<span class="k">for</span> <span class="n">possible_chains</span><span class="p">,</span><span class="n">values</span> <span class="ow">in</span> <span class="n">chain_call</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>			
					<span class="k">if</span> <span class="n">locus</span> <span class="ow">in</span> <span class="n">values</span><span class="p">:</span>
						<span class="n">chain</span> <span class="o">=</span> <span class="n">possible_chains</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
						<span class="k">break</span>		
			
															
			<span class="c">#chain could not be determind with provided locus </span>
			<span class="k">if</span> <span class="n">chain</span> <span class="o">==</span> <span class="s">&#39;N/A&#39;</span><span class="p">:</span>		
				<span class="n">this_chain</span><span class="o">=</span><span class="s">&#39;&#39;</span>
				<span class="n">invalid_chain</span><span class="o">+=</span><span class="mi">1</span>
				<span class="k">print</span> <span class="s">&#39;Error type 1&#39;</span> <span class="c">#isotype is mislabeled</span>
				<span class="k">print</span> <span class="n">my_line</span>			
				<span class="k">continue</span>
			
			<span class="c">#this is the recombination type of the current sequence </span>
			<span class="n">rtype</span><span class="o">=</span><span class="n">chain</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
									
			<span class="c">#unproductive</span>
			<span class="k">if</span> <span class="n">productivity_function_call</span><span class="p">(</span><span class="n">my_line</span><span class="p">)</span><span class="o">==</span><span class="bp">False</span><span class="p">:</span>																									
				<span class="c">#write to temperoary file </span>
				<span class="n">not_productive</span><span class="o">+=</span><span class="mi">1</span>
				<span class="k">continue</span>
								
			<span class="n">CDR3</span> <span class="o">=</span> <span class="n">my_line</span><span class="p">[</span><span class="n">cdr3_field</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
			<span class="n">Len</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">CDR3</span><span class="p">)</span><span class="o">/</span><span class="mi">3</span>	
						
			<span class="c">#get SHM info </span>
			<span class="k">if</span> <span class="n">my_line</span><span class="p">[</span><span class="n">mut_field</span><span class="p">]:</span>
				<span class="k">try</span><span class="p">:</span>
					<span class="n">Mut</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">my_line</span><span class="p">[</span><span class="n">mut_field</span><span class="p">]),</span><span class="mi">3</span><span class="p">)</span>
				<span class="k">except</span><span class="p">:</span>
					<span class="k">if</span> <span class="s">&#39;(&#39;</span> <span class="ow">in</span> <span class="n">my_line</span><span class="p">[</span><span class="n">mut_field</span><span class="p">]:</span>
						<span class="k">try</span><span class="p">:</span>
							<span class="n">Mut</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">my_line</span><span class="p">[</span><span class="n">mut_field</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()),</span><span class="mi">3</span><span class="p">)</span>
						<span class="k">except</span><span class="p">:</span>
							<span class="n">Mut</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span>
					<span class="k">else</span><span class="p">:</span>				
						<span class="n">Mut</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span>		
			<span class="k">else</span><span class="p">:</span>
				<span class="n">Mut</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span>
			
			<span class="c">#get isotype </span>
			<span class="n">isotype</span> <span class="o">=</span> <span class="n">my_line</span><span class="p">[</span><span class="n">isotype_field</span><span class="p">]</span>			
			<span class="n">seq</span><span class="o">=</span><span class="n">my_line</span><span class="p">[</span><span class="n">seq_field</span><span class="p">]</span>						
			<span class="n">v_scores</span> <span class="o">=</span> <span class="n">my_line</span><span class="p">[</span><span class="n">vgene_scores</span><span class="p">]</span>
			<span class="n">d_scores</span> <span class="o">=</span> <span class="n">my_line</span><span class="p">[</span><span class="n">dgene_scores</span><span class="p">]</span>
			<span class="n">j_scores</span> <span class="o">=</span> <span class="n">my_line</span><span class="p">[</span><span class="n">jgene_scores</span><span class="p">]</span>
			
			<span class="n">passed_filter</span><span class="o">+=</span><span class="mi">1</span>
						
			<span class="c">#This specific sequence passed all filters. Write this sequence to a file that can be sorted afterwards by its &#39;full code&#39; afterwards</span>
			<span class="n">temp_array</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;&#39;</span><span class="p">]</span><span class="o">*</span><span class="p">(</span><span class="n">num_elem_stored</span><span class="p">)</span>			
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;CDR3_SEQ&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">CDR3</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;VGENE&#39;</span><span class="p">]]</span><span class="o">=</span><span class="n">Vgene</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;DGENE&#39;</span><span class="p">]]</span><span class="o">=</span><span class="n">Dgene</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;JGENE&#39;</span><span class="p">]]</span><span class="o">=</span><span class="n">Jgene</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;LOCUS&#39;</span><span class="p">]]</span><span class="o">=</span><span class="n">locus</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;MUT&#39;</span><span class="p">]]</span><span class="o">=</span><span class="n">Mut</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;CDR3_LEN&#39;</span><span class="p">]]</span><span class="o">=</span><span class="n">Len</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;AB_SEQ&#39;</span><span class="p">]]</span><span class="o">=</span><span class="n">my_line</span><span class="p">[</span><span class="n">full_nt_field</span><span class="p">]</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;CHAIN_CALL&#39;</span><span class="p">]]</span><span class="o">=</span><span class="n">chain</span> 		
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;ISOTYPE&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">isotype</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;VSCORES&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">v_scores</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;DSCORES&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">d_scores</span>
			<span class="n">temp_array</span><span class="p">[</span><span class="n">ab_field_loc</span><span class="p">[</span><span class="s">&#39;JSCORES&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">j_scores</span>			
			<span class="n">temp_array</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fullcode</span><span class="p">)</span>			
			
			<span class="c">#now store all of the importat fields we want to use for pairing to the temp file defined by temp_seq_data</span>
			<span class="k">if</span> <span class="n">rtype</span><span class="o">==</span><span class="s">&#39;VDJ&#39;</span><span class="p">:</span>
				<span class="n">num_vdj</span> <span class="o">+=</span><span class="mi">1</span>
				<span class="n">temp_output_files_vdj</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">temp_array</span><span class="p">])</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
			<span class="k">elif</span> <span class="n">rtype</span><span class="o">==</span><span class="s">&#39;VJ&#39;</span><span class="p">:</span>
				<span class="n">num_vj</span><span class="o">+=</span><span class="mi">1</span>
				<span class="k">if</span>  <span class="n">include_vl_analysis</span><span class="p">:</span>
					<span class="n">temp_output_files_vj</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">temp_array</span><span class="p">])</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>						
				
		<span class="k">if</span> <span class="n">num_eof</span><span class="o">==</span><span class="nb">len</span><span class="p">(</span><span class="n">list_of_files</span><span class="p">):</span><span class="c">#all files have been read through</span>
			<span class="n">at_least_one_file_open</span><span class="o">=</span><span class="bp">False</span> 	
	
	<span class="c">#temp_seq_data.close()</span>
	
	<span class="n">temp_output_files_vdj</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
	<span class="k">if</span> <span class="n">temp_output_files_vj</span><span class="p">:</span>
		<span class="n">temp_output_files_vj</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
	
			
	<span class="c">#ok the temp file was made. it only consits of sequences which passed the filters above. We now need to sort this file by the MISEQ id and group sequences into proper VH-VL pairs </span>
	<span class="p">[</span><span class="n">num_unique_cdr3</span><span class="p">]</span> <span class="o">=</span> <span class="n">parse_sorted_paired_file</span><span class="p">(</span><span class="n">pairing_temp_file</span><span class="o">+</span><span class="s">&#39;.vdj.txt&#39;</span><span class="p">,</span><span class="s">&#39;vdj&#39;</span><span class="p">)</span>
	<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">usearch_cluster_file</span><span class="p">):</span>		
		<span class="n">num_unique_cdr3_pair_above1</span> <span class="o">=</span> <span class="n">useful</span><span class="o">.</span><span class="n">file_line_count</span><span class="p">(</span><span class="n">usearch_cluster_file</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">num_unique_cdr3_pair_above1</span> <span class="o">=</span> <span class="mi">0</span>
	<span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">pairing_temp_file</span><span class="o">+</span><span class="s">&#39;.vdj.txt&#39;</span><span class="p">)</span>
	<span class="c">#os.remove(pairing_temp_file)				</span>
				
	<span class="k">print</span> <span class="s">&#39;Summary: &#39;</span>
	<span class="k">print</span> <span class="s">&#39;Parsed through {0} sequences&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">total_seqs</span><span class="p">))</span>
	<span class="k">print</span> <span class="s">&#39;{0} ({1}%) sequences did not pass filters: &#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">total_seqs</span><span class="o">-</span><span class="n">passed_filter</span><span class="p">),</span><span class="nb">str</span><span class="p">(</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span>  <span class="p">(</span><span class="mi">100</span><span class="o">*</span><span class="p">(</span><span class="n">total_seqs</span><span class="o">-</span><span class="n">passed_filter</span><span class="p">))</span><span class="o">/</span><span class="n">total_seqs</span><span class="p">),</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">total_seqs</span><span class="o">&gt;</span><span class="mi">0</span> <span class="k">else</span> <span class="s">&#39;0&#39;</span> <span class="p">)</span>
	<span class="c">#print &#39;		{0} sequences did not have an antibody sequence&#39;.format(str(no_result))</span>
	<span class="k">print</span> <span class="s">&#39;		{0} sequences did not have a cdr3&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">no_cdr3_error</span><span class="p">))</span>
	<span class="k">print</span> <span class="s">&#39;		{0} sequences were not productive&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">not_productive</span><span class="p">))</span>
	<span class="k">print</span> <span class="s">&#39;		{0} sequences had an unidentifiable chain type&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">invalid_chain</span><span class="p">))</span>

	<span class="k">print</span> <span class="s">&#39;{0} VDJ sequences&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num_vdj</span><span class="p">))</span>
	<span class="k">print</span> <span class="s">&#39;{0} VJ sequences&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num_vj</span><span class="p">))</span>
		
	<span class="k">print</span> <span class="s">&#39;CDRH3 sequences were collapsed into {0} sequences containing unique CDRH3 sequences: &#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num_unique_cdr3</span><span class="p">))</span>
	<span class="k">print</span> <span class="s">&#39;{0} ({1}%) of these unique pairs were observed more than once&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">num_unique_cdr3_pair_above1</span><span class="p">,</span><span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="mi">100</span><span class="o">*</span><span class="n">num_unique_cdr3_pair_above1</span><span class="o">/</span><span class="n">num_unique_cdr3</span><span class="p">),</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">num_unique_cdr3</span><span class="o">&gt;</span><span class="mi">0</span> <span class="k">else</span> <span class="s">&#39;0&#39;</span><span class="p">)</span>
		   
	<span class="k">return</span> <span class="p">[</span><span class="n">total_seqs</span><span class="p">,</span><span class="n">passed_filter</span><span class="p">,</span><span class="n">no_result</span><span class="p">,</span><span class="n">no_cdr3_error</span><span class="p">,</span><span class="n">not_productive</span><span class="p">,</span><span class="n">invalid_chain</span><span class="p">,</span><span class="n">num_unique_cdr3</span><span class="p">,</span><span class="n">num_unique_cdr3_pair_above1</span><span class="p">,</span><span class="n">num_vdj</span><span class="p">,</span><span class="n">num_vj</span><span class="p">]</span>
	
<div class="viewcode-block" id="Compile_Clusters"><a class="viewcode-back" href="../VHVLPairing.html#immunogrep_singlechain_cdr3_cluster.Compile_Clusters">[docs]</a><span class="k">def</span> <span class="nf">Compile_Clusters</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">,</span><span class="n">cluster_val</span><span class="p">):</span>	
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	</span>
<span class="sd">		Generates a compiled output file summarizing which CDRH3 clusters were made. </span>
<span class="sd">		Reports the cluster size and extra features such as diversity and dominance to determine whether clustering makes sense for H-L pair </span>
<span class="sd">	</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="k">global</span> <span class="n">clustered_final_output_file</span>
	
	<span class="n">clustered_final_output</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">clustered_final_output_file</span><span class="o">+</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">cluster_val</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;.txt&#39;</span><span class="p">,</span><span class="s">&#39;wb&#39;</span><span class="p">)</span>		
	<span class="n">header_output</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;Total Counts&#39;</span><span class="p">,</span><span class="s">&#39;Indiv_Count&#39;</span><span class="p">,</span><span class="s">&#39;Dominance&#39;</span><span class="p">,</span> <span class="s">&#39;Confidence&#39;</span><span class="p">,</span> <span class="s">&#39;Diversity (Shannon)&#39;</span><span class="p">,</span><span class="s">&#39;Diversity (Simpsons)&#39;</span><span class="p">,</span> <span class="s">&#39;Cluster size&#39;</span><span class="p">,</span> <span class="s">&#39;CDRH3 AA&#39;</span><span class="p">,</span> <span class="s">&#39;CDRL3 AA&#39;</span><span class="p">,</span> <span class="s">&#39;CDRH3&#39;</span><span class="p">,</span><span class="s">&#39;CDRL3&#39;</span><span class="p">,</span><span class="s">&#39;VHgene&#39;</span><span class="p">,</span><span class="s">&#39;DHgene&#39;</span><span class="p">,</span><span class="s">&#39;JHgene&#39;</span><span class="p">,</span><span class="s">&#39;VLgene&#39;</span><span class="p">,</span><span class="s">&#39;JLgene&#39;</span><span class="p">,</span><span class="s">&#39;IgH&#39;</span><span class="p">,</span><span class="s">&#39;IgL&#39;</span><span class="p">,</span><span class="s">&#39;Hmut_avg&#39;</span><span class="p">,</span><span class="s">&#39;Hmut_std&#39;</span><span class="p">,</span><span class="s">&#39;L_mut_avg&#39;</span><span class="p">,</span><span class="s">&#39;Lmut_std&#39;</span><span class="p">,</span><span class="s">&#39;HLen&#39;</span><span class="p">,</span><span class="s">&#39;LLen&#39;</span><span class="p">,</span><span class="s">&#39;Hisotype&#39;</span><span class="p">,</span><span class="s">&#39;Lisotype&#39;</span><span class="p">]</span>
	<span class="n">clustered_final_output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">header_output</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
	
	<span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">clustered_dict_raw</span><span class="o">.</span><span class="n">get</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>	
		<span class="n">best_hit</span> <span class="o">=</span> <span class="n">clustered_dict_raw</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">)</span>
		<span class="c">#total_counts found from all CDR3s in a single cluster </span>
		<span class="n">total_counts</span><span class="o">=</span><span class="nb">float</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
				
		<span class="n">indiv_count</span><span class="o">=</span><span class="nb">float</span><span class="p">(</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="c">#total count for the BESt/highest populated CDR3 in the cluster/largest size</span>
		<span class="n">CDRH3_trans</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
		<span class="n">CDRL3_trans</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
		<span class="n">CDRH3</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
		<span class="n">CDRL3</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
		<span class="n">VHgene</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
		<span class="n">DHgene</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
		<span class="n">JHgene</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>		
		<span class="n">VLgene</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
		<span class="n">JLgene</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
		<span class="n">IgH</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
		<span class="n">IgL</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span>
						
		
		<span class="n">HLen</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">16</span><span class="p">]</span>
		<span class="n">LLen</span><span class="o">=</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">17</span><span class="p">]</span>
		
		<span class="n">HIso</span> <span class="o">=</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">18</span><span class="p">]</span>
		<span class="n">LIso</span> <span class="o">=</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">19</span><span class="p">]</span>
		
		<span class="k">if</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">20</span><span class="p">]:</span>
			<span class="n">Hmut_sum</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">20</span><span class="p">])</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">Hmut_sum</span> <span class="o">=</span> <span class="mi">0</span> 
		<span class="k">if</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">21</span><span class="p">]:</span>
			<span class="n">Hmut_sum_sq</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">21</span><span class="p">])</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">Hmut_sum_sq</span><span class="o">=</span><span class="mi">0</span>
		<span class="k">if</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">22</span><span class="p">]:</span>
			<span class="n">Hmut_count</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">22</span><span class="p">])</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">Hmut_count</span> <span class="o">=</span> <span class="mi">0</span>  
		
		<span class="k">if</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">23</span><span class="p">]:</span>
			<span class="n">Lmut_sum</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">23</span><span class="p">])</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">Lmut_sum</span> <span class="o">=</span> <span class="mi">0</span>
		<span class="k">if</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">24</span><span class="p">]:</span>
			<span class="n">Lmut_sum_sq</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">24</span><span class="p">])</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">Lmut_sum_sq</span> <span class="o">=</span> <span class="mi">0</span>
		
		<span class="k">if</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">25</span><span class="p">]:</span>
			<span class="n">Lmut_count</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">best_hit</span><span class="p">[</span><span class="mi">25</span><span class="p">])</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">Lmut_count</span> <span class="o">=</span> <span class="mi">0</span> 
						
		<span class="n">dominance</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">indiv_count</span><span class="o">*</span><span class="mi">100</span><span class="o">/</span><span class="n">total_counts</span><span class="p">)</span> <span class="c">#what percent of the TOP cdr3 contributes to the TOTAL count </span>
		
		<span class="n">freq</span> <span class="o">=</span> <span class="n">indiv_count</span><span class="o">/</span><span class="n">total_counts</span>
		<span class="n">diversity</span> <span class="o">=</span> <span class="n">freq</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">freq</span><span class="p">)</span>
		<span class="n">diversity_v2</span> <span class="o">=</span> <span class="nb">pow</span><span class="p">(</span><span class="n">freq</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
		
		<span class="n">cluster_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">[</span><span class="n">key</span><span class="p">])</span><span class="o">-</span><span class="mi">1</span>
		
		<span class="k">if</span> <span class="n">cluster_size</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
			<span class="n">indiv_count2</span><span class="o">=</span><span class="nb">float</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="c">#this is the counts for the second best CDR3</span>
			<span class="n">confidence</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mi">100</span><span class="o">-</span><span class="mi">100</span><span class="o">*</span><span class="n">indiv_count2</span><span class="o">/</span><span class="n">indiv_count</span><span class="p">)</span> <span class="c">#confidence =&gt; ratio of best cdr3 count/2nd best cdr3 count									</span>
			<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">[</span><span class="n">key</span><span class="p">])):</span>
				<span class="n">other_hits</span> <span class="o">=</span> <span class="n">clustered_dict_raw</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">)</span>
				<span class="k">if</span> <span class="n">other_hits</span><span class="p">[</span><span class="mi">20</span><span class="p">]:</span>
					<span class="n">Hmut_sum</span><span class="o">+=</span><span class="nb">float</span><span class="p">(</span><span class="n">other_hits</span><span class="p">[</span><span class="mi">20</span><span class="p">])</span>
				<span class="k">if</span> <span class="n">other_hits</span><span class="p">[</span><span class="mi">21</span><span class="p">]:</span>
					<span class="n">Hmut_sum_sq</span><span class="o">+=</span><span class="nb">float</span><span class="p">(</span><span class="n">other_hits</span><span class="p">[</span><span class="mi">21</span><span class="p">])</span>
				<span class="k">if</span> <span class="n">other_hits</span><span class="p">[</span><span class="mi">22</span><span class="p">]:</span>
					<span class="n">Hmut_count</span><span class="o">+=</span><span class="nb">float</span><span class="p">(</span><span class="n">other_hits</span><span class="p">[</span><span class="mi">22</span><span class="p">])</span>
				<span class="k">if</span> <span class="n">other_hits</span><span class="p">[</span><span class="mi">23</span><span class="p">]:</span>
					<span class="n">Lmut_sum</span><span class="o">+=</span><span class="nb">float</span><span class="p">(</span><span class="n">other_hits</span><span class="p">[</span><span class="mi">23</span><span class="p">])</span>
				<span class="k">if</span> <span class="n">other_hits</span><span class="p">[</span><span class="mi">24</span><span class="p">]:</span>
					<span class="n">Lmut_sum_sq</span><span class="o">+=</span><span class="nb">float</span><span class="p">(</span><span class="n">other_hits</span><span class="p">[</span><span class="mi">24</span><span class="p">])</span>
				<span class="k">if</span> <span class="n">other_hits</span><span class="p">[</span><span class="mi">25</span><span class="p">]:</span>
					<span class="n">Lmut_count</span><span class="o">+=</span><span class="nb">float</span><span class="p">(</span><span class="n">other_hits</span><span class="p">[</span><span class="mi">25</span><span class="p">])</span>
				<span class="n">freq</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">other_hits</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">/</span><span class="n">total_counts</span>
				<span class="n">diversity</span><span class="o">+=</span><span class="n">freq</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">freq</span><span class="p">)</span>
				<span class="n">diversity_v2</span><span class="o">+=</span><span class="nb">pow</span><span class="p">(</span><span class="n">freq</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
			<span class="n">diversity</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="o">*</span><span class="n">diversity</span><span class="p">)</span>
			<span class="n">diversity_v2</span> <span class="o">=</span> <span class="mi">1</span><span class="o">/</span><span class="n">diversity_v2</span>			
			<span class="k">if</span> <span class="n">Hmut_count</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
				<span class="n">Hmut_avg</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">Hmut_sum</span><span class="o">/</span><span class="n">Hmut_count</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">Hmut_avg</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span>			
			<span class="k">if</span> <span class="n">Hmut_count</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
				<span class="n">h1</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">Hmut_sum_sq</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
				<span class="n">h2</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">pow</span><span class="p">(</span><span class="n">Hmut_sum</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span><span class="o">/</span><span class="n">Hmut_count</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>				
				<span class="n">Hmut_var</span> <span class="o">=</span>  <span class="nb">round</span><span class="p">(</span><span class="nb">pow</span><span class="p">(</span> <span class="p">(</span><span class="n">h1</span><span class="o">-</span><span class="n">h2</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">Hmut_count</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span><span class="mf">0.5</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">Hmut_var</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span>			
			<span class="k">if</span> <span class="n">Lmut_count</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
				<span class="n">Lmut_avg</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">Lmut_sum</span><span class="o">/</span><span class="n">Lmut_count</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>			
			<span class="k">else</span><span class="p">:</span>
				<span class="n">Lmut_avg</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span>				
			<span class="c">#shm variance -&gt; sum of squares formulat =&gt; sum(vals^2)-(sum(vals)^2)/counts)</span>
			<span class="k">if</span> <span class="n">Lmut_count</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
				<span class="n">l1</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">Lmut_sum_sq</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
				<span class="n">l2</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">pow</span><span class="p">(</span><span class="n">Lmut_sum</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span><span class="o">/</span><span class="n">Lmut_count</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>				
				<span class="n">Lmut_var</span> <span class="o">=</span>  <span class="nb">round</span><span class="p">(</span><span class="nb">pow</span><span class="p">(</span> <span class="p">(</span><span class="n">l1</span><span class="o">-</span><span class="n">l2</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">Lmut_count</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span><span class="mf">0.5</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>				
			<span class="k">else</span><span class="p">:</span>
				<span class="n">Lmut_var</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span>			
		<span class="k">else</span><span class="p">:</span>				
			<span class="n">Hmut_avg</span> <span class="o">=</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">12</span><span class="p">]</span>
			<span class="n">Hmut_var</span> <span class="o">=</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">13</span><span class="p">]</span>
			<span class="n">Lmut_avg</span> <span class="o">=</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">14</span><span class="p">]</span>
			<span class="n">Lmut_var</span> <span class="o">=</span> <span class="n">best_hit</span><span class="p">[</span><span class="mi">15</span><span class="p">]</span>
			<span class="n">confidence</span><span class="o">=</span><span class="mi">100</span>
			<span class="n">diversity</span> <span class="o">=</span> <span class="mi">1</span>
			<span class="n">diversity_v2</span> <span class="o">=</span> <span class="mi">1</span>
						
		<span class="k">try</span><span class="p">:</span> <span class="n">CDRH3_trans</span><span class="o">=</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">CDRH3</span><span class="p">,</span><span class="n">generic_dna</span><span class="p">))</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
		<span class="k">except</span><span class="p">:</span> <span class="n">CDRH3</span><span class="o">=</span> <span class="s">&#39;translation error&#39;</span>
		<span class="k">try</span><span class="p">:</span> <span class="n">CDRL3_trans</span><span class="o">=</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">CDRL3</span><span class="p">,</span><span class="n">generic_dna</span><span class="p">))</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
		<span class="k">except</span><span class="p">:</span> <span class="n">CDRL3</span><span class="o">=</span> <span class="s">&#39;translation error&#39;</span>
		
		<span class="n">results</span> <span class="o">=</span> <span class="p">[</span><span class="n">total_counts</span><span class="p">,</span><span class="n">indiv_count</span><span class="p">,</span><span class="n">dominance</span><span class="p">,</span><span class="n">confidence</span><span class="p">,</span><span class="n">diversity</span><span class="p">,</span><span class="n">diversity_v2</span><span class="p">,</span><span class="n">cluster_size</span><span class="p">,</span><span class="n">CDRH3_trans</span><span class="p">,</span><span class="n">CDRL3_trans</span><span class="p">,</span><span class="n">CDRH3</span><span class="p">,</span><span class="n">CDRL3</span><span class="p">,</span><span class="n">VHgene</span><span class="p">,</span><span class="n">DHgene</span><span class="p">,</span><span class="n">JHgene</span><span class="p">,</span><span class="n">VLgene</span><span class="p">,</span><span class="n">JLgene</span><span class="p">,</span><span class="n">IgH</span><span class="p">,</span><span class="n">IgL</span><span class="p">,</span><span class="n">Hmut_avg</span><span class="p">,</span><span class="n">Hmut_var</span><span class="p">,</span><span class="n">Lmut_avg</span><span class="p">,</span><span class="n">Lmut_var</span><span class="p">,</span><span class="n">HLen</span><span class="p">,</span><span class="n">LLen</span><span class="p">,</span><span class="n">HIso</span><span class="p">,</span> <span class="n">LIso</span><span class="p">]</span>
		<span class="n">results</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">results</span><span class="p">]</span>		
		<span class="n">clustered_final_output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>	

				</div>
<span class="k">def</span> <span class="nf">WriteSummaryFile</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">,</span><span class="n">total_seqs</span><span class="p">,</span><span class="n">passed_filter</span><span class="p">,</span><span class="n">no_result</span><span class="p">,</span><span class="n">no_cdr3_error</span><span class="p">,</span><span class="n">not_productive</span><span class="p">,</span><span class="n">invalid_chain</span><span class="p">,</span><span class="n">num_unique_cdrh3_l3_pair</span><span class="p">,</span><span class="n">num_unique_cdrh3_l3_pair_above1</span><span class="p">,</span><span class="n">cluster</span><span class="p">,</span><span class="n">num_clustered</span><span class="p">,</span><span class="n">num_vdj</span><span class="p">,</span><span class="n">num_vj</span><span class="p">):</span>
	<span class="c">#MAKE THIS SPECIAL IF ITS IMGT FILES (MERGE FILES INTO ONE GROUP NAME)	</span>
	<span class="n">file_path_summary</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">+</span><span class="n">f</span><span class="o">+</span><span class="s">&#39;,&#39;</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">annotated_file_paths</span><span class="p">]</span>

	<span class="k">global</span> <span class="n">summary_file</span>
	<span class="n">summary</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span>	<span class="n">summary_file</span><span class="p">,</span><span class="s">&#39;wb&#39;</span><span class="p">)</span>
	<span class="c">#Write useful information in summary file	</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Data generated using </span><span class="si">%s</span><span class="s"> on: </span><span class="si">%s</span><span class="se">\n</span><span class="s">&#39;</span> <span class="o">%</span><span class="p">(</span><span class="n">pairing_version</span><span class="p">,</span><span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()))</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;***************************************************</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;***************   Summary Report  *****************</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;***************************************************</span><span class="se">\n\n\n</span><span class="s">&quot;</span><span class="p">)</span>
	
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;The following input files were used to run pairing analysis:</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_path_summary</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n\n\n</span><span class="s">&#39;</span><span class="p">)</span>
	
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Parsed through {0} sequences</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">total_seqs</span><span class="p">)))</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{0} ({1}%) sequences did not pass filters:</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">total_seqs</span><span class="o">-</span><span class="n">passed_filter</span><span class="p">),</span><span class="nb">str</span><span class="p">(</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span>  <span class="p">(</span><span class="mi">100</span><span class="o">*</span><span class="p">(</span><span class="n">total_seqs</span><span class="o">-</span><span class="n">passed_filter</span><span class="p">))</span><span class="o">/</span><span class="n">total_seqs</span><span class="p">),</span><span class="mi">1</span><span class="p">)))</span> <span class="k">if</span> <span class="n">total_seqs</span><span class="o">&gt;</span><span class="mi">0</span> <span class="k">else</span> <span class="s">&#39;0&#39;</span><span class="p">)</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">{0} sequences did not have an antibody sequence</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">no_result</span><span class="p">)))</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">{0} sequences did not have a cdr3</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">no_cdr3_error</span><span class="p">)))</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">{0} sequences were not productive</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">not_productive</span><span class="p">)))</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">{0} sequences had an unidentifiable chain type</span><span class="se">\n\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">invalid_chain</span><span class="p">)))</span>

	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{0} Sequences had were identified as a VDJ recombination event</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num_vdj</span><span class="p">)))</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{0} Sequences had were identified as a VJ recombination event</span><span class="se">\n\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num_vj</span><span class="p">)))</span>
		
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Sequence-pairs were collapsed into {0} sequences containing unique CDR3s</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num_unique_cdrh3_l3_pair</span><span class="p">)))</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{0} ({1}%) of these unique CDRH3 were observed more than once</span><span class="se">\n\n\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">num_unique_cdrh3_l3_pair_above1</span><span class="p">,</span><span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="mi">100</span><span class="o">*</span><span class="n">num_unique_cdrh3_l3_pair_above1</span><span class="o">/</span><span class="n">num_unique_cdrh3_l3_pair</span><span class="p">),</span><span class="mi">1</span><span class="p">)))</span> <span class="k">if</span> <span class="n">num_unique_cdrh3_l3_pair</span><span class="o">&gt;</span><span class="mi">0</span> <span class="k">else</span> <span class="s">&#39;0&#39;</span><span class="p">)</span>
		
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;Number of unique CDRH3 sequences: </span><span class="si">%d</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">num_unique_cdrh3_l3_pair</span><span class="p">))</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;CDRH3 sequences were clustered using the following cluster cutoffs</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">Cluster Cutoff</span><span class="se">\t</span><span class="s">Number of clustered sequences</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
	
	<span class="k">for</span> <span class="n">cpos</span><span class="p">,</span><span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cluster</span><span class="p">):</span>
		<span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="o">*</span><span class="mi">100</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">num_clustered</span><span class="p">[</span><span class="n">cpos</span><span class="p">])</span><span class="o">+</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
	
	<span class="c">#summary.write(&quot;\t Number of clustered sequences to %d%% CDRH3 identity: %d\n&quot; %(cluster*100,num_clustered))</span>
	<span class="n">summary</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>



<div class="viewcode-block" id="RunSingleChainClustering"><a class="viewcode-back" href="../VHVLPairing.html#immunogrep_singlechain_cdr3_cluster.RunSingleChainClustering">[docs]</a><span class="k">def</span> <span class="nf">RunSingleChainClustering</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">,</span><span class="n">analysis_method</span><span class="p">,</span><span class="n">output_folder_path</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">,</span><span class="n">prefix_output_files</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">,</span> <span class="n">annotated_file_formats</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span><span class="n">field_names</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span><span class="n">include_vl_analysis</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span><span class="n">cluster_cutoff</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.96</span><span class="p">,</span><span class="mf">0.96</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span><span class="n">files_from_igrep_database</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span><span class="n">productivity_function</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		</span>
<span class="sd">		Brief Description </span>
<span class="sd">		</span>
<span class="sd">		This function serves as an addition to the current function RunPairing found in immunogrep_gglab_pairing.py file. </span>
<span class="sd">	    This function preforms the same outlined steps and outputs the same information as the pairing program except it only considers the CDR3 of each chain seperately.</span>
<span class="sd">		Whereas RunPairing function will group together sequences by their respective R1-R2 reads (VH-VL pairs), this function will consider each chain independently. </span>
<span class="sd">		Clustering therefore only considers VH only generated libraries and VL only generated libraries.</span>
<span class="sd">		Clustering of CDR3 sequences is again performed using Usearch program, and the output file should match the format of the RunPairing program except excludes VH-VL pairs.</span>
<span class="sd">		</span>
<span class="sd">		General algorithm:</span>
<span class="sd">		</span>
<span class="sd">		The single chain analysis program follows the following steps: </span>
<span class="sd">			Step A: Go through the provided AB annotated files (IMGT, parsed IGBLAST files, etc), and filter out sequences that lack good results </span>
<span class="sd">				1) Loop through all sequences provided in all files </span>
<span class="sd">				2) Only select sequences that have a sequence header. </span>
<span class="sd">				3) Filter out sequences that do not have a cdr3 sequence </span>
<span class="sd">				4) Filter out sequences that are deemed &#39;unproductive&#39; using our Productivity Rules, or a custom productivity rule provided by user 						</span>
<span class="sd">				5) Seperate VDJ annotated sequences from VJ annoated sequences into seperate files</span>
<span class="sd">				6) The user provided parameter (include_vl_analysis) which determine whether to cluster both the VDJ and VJ files. By default we only consider VDJ data.</span>

<span class="sd">			Step B: Group together CDRH3 sequences for each VDJ and/or VJ file</span>
<span class="sd">				1) We group together sequences with identical CDRH3 sequences from the filtered sequences in Step B</span>
<span class="sd">				2) For each CDRH3-CDRL3 group, we calculate the following as extra information for the group: </span>
<span class="sd">					a) average SHM (if an SHM field is provided) and standard deviation SHM </span>
<span class="sd">					b) mode of V,D, and J gene calls (if V, D, or J genes are provided). Mode of gene calls are based on top genes only (genes whose alignment scores are equal to the top alignment score)</span>
<span class="sd">					c) All observed unique isotypes (if isotypes are provided)</span>
<span class="sd">			</span>
<span class="sd">			Step D: Cluster CDRH3-CDRL3 pairs </span>
<span class="sd">				1) Unique CDRH3-CDRL3 paired sequences are sorted by their total counts (number of sequences containing respective CDRH3-CDRL3 pair)</span>
<span class="sd">				2) The results from each CDRH3-CDRL3, pair whose counts are found &gt; 1 time, are exported as a FASTA file. </span>
<span class="sd">					a) The sequence header contains all of the data generated/information we are interested in for the H3-L3 pair. </span>
<span class="sd">					b) The sequence value for the FASTA is the nucleotide sequence of the CDRH3 (heavy chain) sequence only </span>
<span class="sd">				3) This FASTA file is used as an input file to the program USEARCH VERSION 7. </span>
<span class="sd">				4) We run the Usearch program for each cluster defined in the cluster_cutoff parameter defined in this function</span>
<span class="sd">			</span>
<span class="sd">			Step E: Summarize the output from each USEARCH cluster file </span>
<span class="sd">				1) For EACH cluster cutoff defined, USEARCH creates an output .uc file </span>
<span class="sd">				2) We parse this file and generate a finalized output file based on the CDRH3 seed/cluster</span>
<span class="sd">				3) For each seed we calculate the following</span>
<span class="sd">					a) The average SHM for all sequences which belong in the seed </span>
<span class="sd">					b) The V,D, and J call of the SEED sequence </span>
<span class="sd">					c) The Isotype call of the SEED sequence</span>
<span class="sd">					d) A &#39;dominance&#39; calculation =&gt; </span>
<span class="sd">						100*Unique CDR3 Count of the seed sequence/Unique CDR3 Count of the next largest CDR3 group in the cluster</span>
<span class="sd">					e) A &#39;confidence&#39; calculation =&gt; </span>
<span class="sd">						100*Unique CDR3 Count of the seed sequence/The total counts of all sequences in cluster</span>
<span class="sd">					f) Diversity of all members in seed using </span>
<span class="sd">						i. shannon entropy </span>
<span class="sd">						ii. simpsons index </span>
<span class="sd">								</span>
<span class="sd">		General usage:</span>
<span class="sd">		</span>
<span class="sd">		When calling this function, the most important feature for this program to run correctly is defining the type of file(s) you have submitted. While there is no uniform file type/format, we do need to know which field names in the file correspond to field names we use in the pairing program. </span>
<span class="sd">		The variable, required_field_names, stores the fields we need to perform the analysis. From these fields, only the following fields are absolutely necessary (although providing all fields are ideal):</span>
<span class="sd">			1) seq_header</span>
<span class="sd">			2) cdr3_nt</span>
<span class="sd">					</span>
<span class="sd">		We already know which field names are required for files that were generated using our common annotation methods: IMGT, IGFFT (in house program), MIXCR, IGBLAST. Therefore, if you are pairing files from one of these programs AND you have not modified the output of the files created by our python wrapper functions, there is no need to define the fields. </span>
<span class="sd">		However, if the file being provided is a custom file or an annotation file we do not know about, then you need to define custom field names in the input variable: field_names</span>
<span class="sd">										</span>
<span class="sd">		Example usage for pairing IMGT files: 			</span>
<span class="sd">			list_of_imgt_files = a list of strings providing the location of all IMGT generated files for pairing. You do not need to select a specific IMGT file or differente the 11 files by experiment, we handle that in the function</span>
<span class="sd">			list_of_imgt_files = [1_summary_file1.txt,5_summary_file1.txt,1_summary_file2.txt,5_summary_file2.txt] (etc  etc for the files to pair)</span>
<span class="sd">			RunPairing(annotated_file_paths = list_of_imgt_files,analysis_method = IMGT, output_folder_path=&#39;paired_seq_data.txt,prefix_output_files=&#39;MYPAIRINGEXP&#39;,annotated_file_formats=&#39;IMGT&#39;,field_names=None,cluster_cutoff=[0.85,0.96,0.01])</span>
<span class="sd">		</span>
<span class="sd">		Example use for pairing MIXCR files:</span>
<span class="sd">			list_of_mixcr_files = a list of strings providing the location of all MIXCR annotation files generated by the function ParseMIXCR</span>
<span class="sd">			list_of_mixcr_files = [file1.mixcr.annotation,file2.mixcr.annotation,file3.mixcr.annotation]</span>
<span class="sd">			RunPairing(annotated_file_paths = list_of_mixcr_files,analysis_method = MIXCR, output_folder_path=&#39;paired_seq_data.txt,prefix_output_files=&#39;MYPAIRINGEXP&#39;,annotated_file_formats=&#39;TAB&#39;,field_names=None,cluster_cutoff=[0.85,0.96,0.01])</span>
<span class="sd">		</span>
<span class="sd">		Example use for any custom file (a file type not generated by the database, parseMIXCR function, parseIgblast function, parseIGFFT function, or IMGT):</span>
<span class="sd">			list_of_files = a list of strings providing the location of some CSV file that we have not seen before </span>
<span class="sd">			In this example, this CSV (not TAB) file only contains CDR3 nucleotide information. So every other field provided will be treated as blank for the program and not considered. In the provided file, the CDR3 field is labeled as CDR3 Nt and the sequence header is labeled as, Miseq header</span>
<span class="sd">			field_mappings = { #key = name of the field we need in the program, value = name of the field in the provided file </span>
<span class="sd">				&#39;cdr3_nt&#39;:	&#39;CDR3 Nt&#39;,</span>
<span class="sd">				&#39;seq_header&#39;:Miseq header&#39;</span>
<span class="sd">			}</span>
<span class="sd">			list_of_files = [fie1.txt,file2.txt,file3.txt]</span>
<span class="sd">			</span>
<span class="sd">			RunPairing(annotated_file_paths = list_of_files,analysis_method = custom, output_folder_path=&#39;paired_seq_data.txt,prefix_output_files=&#39;MYCUSTOMPAIRINGEXP&#39;,annotated_file_formats=&#39;CSV&#39;,field_names=field_mappings,cluster_cutoff=[0.85,0.96,0.01])</span>
<span class="sd">			</span>
<span class="sd">			</span>
<span class="sd">		#REQUIRED FIELDS###</span>
<span class="sd">		#annotated_file_paths = a list of files that you want to run pairing with </span>
<span class="sd">		#analysis_method:</span>
<span class="sd">			#A string defining which annotation program generated the results </span>
<span class="sd">			#IF an annotation program is not defined in this script (variable above) yet, then use &#39;CUSTOM&#39;</span>
<span class="sd">			#note: if analysis_method == &#39;CUSTOM&#39; OR not in the predefined list, then the parameter,field_names, is required </span>
<span class="sd">		</span>
<span class="sd">		#OPTIONAL FIELDS###</span>
<span class="sd">		#annotated_file_formats: </span>
<span class="sd">			#accepts three possible formats: </span>
<span class="sd">				#=&gt; None =&gt; do not define file type, let program guess </span>
<span class="sd">				#=&gt; list of file types =&gt; the file type FOR EACH PROVIDED FILE </span>
<span class="sd">				#=&gt; SINGLE STRING =&gt; THIS MEANS ALL INPUTED FILES ARE OF THE SAME TYPE </span>
<span class="sd">		</span>
<span class="sd">		#field_names: </span>
<span class="sd">			#A dictionary defining which fields in the file correspond to the fields required for this program </span>
<span class="sd">			#IMPORTANT: If analysis_method == CUSTOM then you must define field_names in the file. Program will raise exception otherwise</span>
<span class="sd">			#if provided, the structure of field names is as follows: </span>
<span class="sd">				#keys of variable: &#39;vgene&#39;,&#39;jgene&#39;,&#39;dgene&#39;,&#39;raw_seq_nt&#39;,&#39;cdr3_nt&#39;,&#39;seq_header&#39;,&#39;shm&#39;,&#39;full_len_ab_aa&#39;</span>
<span class="sd">				#values =&gt; for each key,provide the field name in the file that corresponds to the key </span>
<span class="sd">				</span>
<span class="sd">		#output_folder_path =&gt; path of output folder. If empty, then a new folder is created. If folder path does not exist, then creates a new folder using provided folder path</span>
<span class="sd">		#prefix_output_files =&gt; prefix string to use for naming output files</span>
<span class="sd">		#cluster_cutoff =&gt; clustering cutoff for determining clustered pairs  (accepts a single number or a list of 3 numbers)</span>
<span class="sd">			If its a single number, i.e. cluster = 0.96, then only that clustering id cutoff is used </span>
<span class="sd">			If a list is provided, then a range of clusters are attempted defined by the list [min cluster id, max cluster id, step increase of cluster id&#39;s]</span>
<span class="sd">		#files_from_igrep_database =&gt; indicates whether the provided files were downloaded from the database (this is important because files from the database will have the same field names regardless of the annotation type. Therefore, we need to use these files from database</span>
<span class="sd">		</span>
<span class="sd">	&quot;&quot;&quot;</span>						
		
	<span class="n">required_field_names</span> <span class="o">=</span> <span class="n">get_required_field_names</span><span class="p">()</span>	
			
	<span class="c">###########INITIALIZATION#################</span>
	<span class="c">##########################################</span>
	<span class="c">#JUST SET UP INPUT PARAMETERS. GO THROUGH PROVIDED PARAMETERS AND DETERMINE HOW TO RUN PROGRAM#		</span>
	<span class="c">#cluster cutof must be between 0 and 1 </span>
	<span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cluster_cutoff</span><span class="p">,</span><span class="nb">list</span><span class="p">):</span>
		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cluster_cutoff</span><span class="p">)</span><span class="o">!=</span><span class="mi">3</span><span class="p">:</span>
			<span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;The cluster range must be defined as a 3 element list of :[min cluster,max cluster, step increase in cluster]&#39;</span><span class="p">)</span>
		<span class="n">cluster_cutoff</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">FloatRange</span><span class="p">(</span><span class="n">cluster_cutoff</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">cluster_cutoff</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">cluster_cutoff</span><span class="p">[</span><span class="mi">2</span><span class="p">])]</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">if</span> <span class="n">cluster_cutoff</span><span class="o">&lt;=</span><span class="mi">0</span> <span class="ow">or</span> <span class="n">cluster_cutoff</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
			<span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;Cluster cutoff must be a floating number between 0 and 1&#39;</span><span class="p">)</span>			
		<span class="n">cluster_cutoff</span> <span class="o">=</span> <span class="p">[</span><span class="n">cluster_cutoff</span><span class="p">]</span>		
	

	<span class="n">cluster_cutoff</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">cluster_cutoff</span><span class="p">)))]</span>			
	
	<span class="c">#if user just passe in a single string for single file, then make it a list </span>
	<span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">,</span><span class="nb">list</span><span class="p">):</span>
		<span class="n">annotated_file_paths</span> <span class="o">=</span> <span class="p">[</span><span class="n">annotated_file_paths</span><span class="p">]</span>			
	<span class="c">#user passed in the file type for EACH provided file </span>
	<span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">annotated_file_formats</span><span class="p">,</span><span class="nb">list</span><span class="p">):</span>
		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annotated_file_formats</span><span class="p">)</span><span class="o">!=</span><span class="nb">len</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">):</span>
			<span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;The user defined a file format for each file. But the number of file formats does not match the number of input files. If the user wants to use a single format for all files, then simply pass in a string not list defining file format&#39;</span><span class="p">)</span>
		
		<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">annotated_file_formats</span><span class="p">)):</span>
			<span class="n">annotated_file_formats</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">annotated_file_formats</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
			<span class="k">if</span> <span class="n">annotated_file_formats</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowed_file_formats</span><span class="p">:</span>
				<span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;The provided file format, {0}, is not currently supported in this program&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">annotated_file_formats</span><span class="p">))</span>		
	<span class="c">#user just passed in a string. copy this to all a list contaiing same string for all files </span>
	<span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">annotated_file_formats</span><span class="p">,</span><span class="nb">basestring</span><span class="p">):</span>		
		<span class="n">annotated_file_formats</span> <span class="o">=</span> <span class="n">annotated_file_formats</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>		
		<span class="k">if</span> <span class="n">annotated_file_formats</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowed_file_formats</span><span class="p">:</span>
			<span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;The provided file format, {0}, is not currently supported in this program&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">annotated_file_formats</span><span class="p">))</span>	
		<span class="n">annotated_file_formats</span> <span class="o">=</span> <span class="p">[</span><span class="n">annotated_file_formats</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">)</span>	
	<span class="c">#user provided ouput folder </span>
	<span class="k">if</span> <span class="n">output_folder_path</span><span class="p">:</span>		
		<span class="c">#folder path provided does not exist </span>
		<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">output_folder_path</span><span class="p">):</span>			
			<span class="c">#make folder, but ONLY USE BASENAME. FORCE the folder to exist within IGREP project </span>
			<span class="k">if</span> <span class="n">output_folder_path</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">==</span><span class="s">&#39;/&#39;</span><span class="p">:</span>
				<span class="n">output_folder_path</span><span class="o">=</span><span class="n">output_folder_path</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
			<span class="n">parent_dir</span> <span class="o">=</span> <span class="n">filesystem</span><span class="o">.</span><span class="n">ExperimentDirs</span><span class="p">()</span><span class="o">.</span><span class="n">make_dir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">output_folder_path</span><span class="p">))</span><span class="o">+</span><span class="s">&#39;/&#39;</span>
			<span class="k">print</span> <span class="s">&quot;WARNING, OUTPUT DIRECTORY NOT FOUND. CREATED A NEW DIRECTORY FOR OUTPUT CALLED: {0}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">parent_dir</span><span class="p">)</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="c">#this is the output folder </span>
			<span class="n">parent_dir</span> <span class="o">=</span> <span class="n">output_folder_path</span><span class="o">+</span><span class="s">&#39;/&#39;</span> <span class="k">if</span> <span class="n">output_folder_path</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">!=</span><span class="s">&#39;/&#39;</span> <span class="k">else</span> <span class="n">output_folder_path</span>									
	<span class="k">else</span><span class="p">:</span>
		<span class="c">#user did not pass in output folder, so make a new folder</span>
		<span class="n">folder_name</span> <span class="o">=</span> <span class="s">&#39;PAIRING_RESULT_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">())</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39; &#39;</span><span class="p">,</span><span class="s">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">,</span><span class="s">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;-&#39;</span><span class="p">,</span><span class="s">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">,</span><span class="s">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
		<span class="c">#make a new folder</span>
		<span class="n">parent_dir</span> <span class="o">=</span> <span class="n">filesystem</span><span class="o">.</span><span class="n">ExperimentDirs</span><span class="p">()</span><span class="o">.</span><span class="n">make_dir</span><span class="p">(</span><span class="n">folder_name</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;/&#39;</span>		
	
	<span class="c">#replace empty spaces in folder name (can be annoying..)</span>
	<span class="n">parent_dir</span> <span class="o">=</span> <span class="n">parent_dir</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39; &#39;</span><span class="p">,</span><span class="s">&#39;_&#39;</span><span class="p">)</span>
	<span class="n">prefix</span> <span class="o">=</span> <span class="n">parent_dir</span><span class="o">+</span><span class="n">prefix_output_files</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39; &#39;</span><span class="p">,</span><span class="s">&#39;_&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">prefix_output_files</span> <span class="k">else</span> <span class="n">parent_dir</span><span class="o">+</span><span class="s">&#39;VH_VL_PAIRING&#39;</span>	
	<span class="c">#make sure provided analysis settings are correct/make sure we have all the data we need for pairing </span>
	<span class="n">analysis_method</span><span class="o">=</span><span class="n">analysis_method</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
	<span class="k">if</span> <span class="n">analysis_method</span> <span class="o">==</span> <span class="s">&#39;CUSTOM&#39;</span> <span class="ow">or</span> <span class="n">analysis_method</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">supported_analyses</span><span class="p">:</span>
		<span class="k">if</span> <span class="ow">not</span> <span class="n">field_names</span><span class="p">:</span>
			<span class="k">raise</span> <span class="n">Excpetion</span><span class="p">(</span><span class="s">&#39;You have defined the following analysis method: {0}. However, we cannot use this file until we know the field names for the following variables: {1}. Please rerun function defining these values as key:value format (key = required field name, value = field name in file)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">analysis_method</span><span class="p">,</span><span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">required_field_names</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>
		<span class="k">for</span> <span class="n">each_req_field</span><span class="p">,</span><span class="n">value</span> <span class="ow">in</span> <span class="n">field_names</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
			<span class="k">if</span> <span class="n">each_req_field</span> <span class="ow">in</span> <span class="n">required_field_names</span><span class="p">:</span>
				<span class="n">required_field_names</span><span class="p">[</span><span class="n">each_req_field</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>							
	<span class="k">else</span><span class="p">:</span>
		<span class="c">#the user did not provide field_names, but thats ok because we already know the analysis type and have it hardcoded in variable above </span>
		<span class="k">if</span> <span class="ow">not</span> <span class="n">field_names</span><span class="p">:</span>
			<span class="k">for</span> <span class="n">each_req_field</span><span class="p">,</span><span class="n">value</span> <span class="ow">in</span> <span class="n">pairing_settings</span><span class="p">[</span><span class="n">analysis_method</span><span class="p">][</span><span class="s">&#39;fields_for_analysis&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
				<span class="k">if</span> <span class="n">each_req_field</span> <span class="ow">in</span> <span class="n">required_field_names</span><span class="p">:</span>
					<span class="n">required_field_names</span><span class="p">[</span><span class="n">each_req_field</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>							
		<span class="c">#the user provided the field name values </span>
		<span class="k">else</span><span class="p">:</span>
			<span class="k">for</span> <span class="n">each_req_field</span><span class="p">,</span><span class="n">value</span> <span class="ow">in</span> <span class="n">field_names</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
				<span class="k">if</span> <span class="n">each_req_field</span> <span class="ow">in</span> <span class="n">required_field_names</span><span class="p">:</span>
					<span class="n">required_field_names</span><span class="p">[</span><span class="n">each_req_field</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>				
		<span class="k">if</span> <span class="n">productivity_function</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span>
			<span class="n">productivity_function</span> <span class="o">=</span> <span class="n">pairing_settings</span><span class="p">[</span><span class="n">analysis_method</span><span class="p">][</span><span class="s">&#39;productivity_function&#39;</span><span class="p">]</span>
		
	<span class="c">#at the very least, both the field for seq header and the field for cdr3 nt must be defined </span>
	<span class="k">if</span> <span class="ow">not</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;seq_header&#39;</span><span class="p">]</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">required_field_names</span><span class="p">[</span><span class="s">&#39;cdr3_nt&#39;</span><span class="p">]:</span>
		<span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;At the very least you must define what field name corresponds to the seq_header and what field name corresponds to the nucleotide cdr3. The following fields were defined by the variable field_names: &#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">field_names</span><span class="p">))</span>
	
	<span class="c">#NOW tell the user whether they are will be missing any fields with the provided file for pairing (therefore they know whether to expect weird results)</span>
	<span class="k">for</span> <span class="n">each_req_field</span><span class="p">,</span><span class="n">value</span> <span class="ow">in</span> <span class="n">required_field_names</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
		<span class="k">if</span> <span class="ow">not</span> <span class="n">value</span><span class="p">:</span>
			<span class="k">print</span> <span class="s">&quot;WARNING, THE FOLLOWING FIELD IS NOT DEFINED: {0}. PAIRING WILL NOT INCLUDE INFORMATION FOR THIS FIELD&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">each_req_field</span><span class="p">)</span>			
	
	<span class="c">#WE PREDICT TWO WAYS TO PROVIDE IMGT FILES: </span>
		<span class="c">#METHOD 1=&gt; PROVIDE THE ORIGINAL 11 IMGT FILES, IN THIS CASE, THEN ANLAYSIS_METHOD == IMGT AND ALSO ANNOTATED_FILE_FORMATS = IMGT OR [IMGT,IMGT,IMGT]</span>
		<span class="c">#METHOD 2=&gt; PROVIDE A SUMMARIZED FILE COMBINING RESULTS FROM IMGT. FOR EXAMPLE DOWNLOADING IMGT RESULTS FROM DATABASE. IN THIS CASE THEN ANALYSIS_METHOD == IMGT BUT!! THE ANNOTATED_FILE_FORMATS DOES  NOT EQUAL IMGT 		</span>
	<span class="c">#SO IF METHOD 1, THEN WE WILL REORGANIZE THE PROVIDED FILES. WE DO THIS BECAUSE WE WANT TO READ MULTIPLE EXPERIMENTS simultaneously (READ MULTIPLE SETS OF IMGT FILES AT THE SAME TIME TO MAINTAIN THE SIZE OF CODEDICT AS SMALL)</span>
	<span class="k">if</span> <span class="n">analysis_method</span><span class="o">==</span><span class="s">&#39;IMGT&#39;</span><span class="p">:</span>
		<span class="k">if</span> <span class="n">annotated_file_formats</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span>			
			<span class="n">annotated_file_formats</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;IMGT&#39;</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">)</span>		
			<span class="n">imgt_groups</span> <span class="o">=</span> <span class="n">readfile</span><span class="o">.</span><span class="n">GroupIMGTFiles</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">)</span>		
			<span class="n">annotated_file_paths</span> <span class="o">=</span> <span class="p">[</span><span class="n">vals</span> <span class="k">for</span> <span class="n">group</span><span class="p">,</span><span class="n">vals</span> <span class="ow">in</span> <span class="n">imgt_groups</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()]</span>			
		<span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">annotated_file_formats</span><span class="p">,</span><span class="nb">list</span><span class="p">):</span>
			<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annotated_file_formats</span><span class="p">)</span><span class="o">!=</span><span class="nb">len</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">):</span>
				<span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;The provided number of file formats does not match the provided files&#39;</span><span class="p">)</span>
			<span class="k">if</span> <span class="n">annotated_file_formats</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;IMGT&#39;</span><span class="p">:</span>
				<span class="n">annotated_file_formats</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;IMGT&#39;</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">)</span>		
				<span class="n">imgt_groups</span> <span class="o">=</span> <span class="n">readfile</span><span class="o">.</span><span class="n">GroupIMGTFiles</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">)</span>		
				<span class="n">annotated_file_paths</span> <span class="o">=</span> <span class="p">[</span><span class="n">vals</span> <span class="k">for</span> <span class="n">group</span><span class="p">,</span><span class="n">vals</span> <span class="ow">in</span> <span class="n">imgt_groups</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()]</span>			
		<span class="k">elif</span> <span class="n">annotated_file_formats</span> <span class="o">==</span> <span class="s">&#39;IMGT&#39;</span><span class="p">:</span>
			<span class="n">annotated_file_formats</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;IMGT&#39;</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">)</span>		
			<span class="n">imgt_groups</span> <span class="o">=</span> <span class="n">readfile</span><span class="o">.</span><span class="n">GroupIMGTFiles</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">)</span>		
			<span class="n">annotated_file_paths</span> <span class="o">=</span> <span class="p">[</span><span class="n">vals</span> <span class="k">for</span> <span class="n">group</span><span class="p">,</span><span class="n">vals</span> <span class="ow">in</span> <span class="n">imgt_groups</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()]</span>			
		<span class="k">else</span><span class="p">:</span>
			<span class="n">annotated_file_formats</span> <span class="o">=</span> <span class="p">[</span><span class="n">annotated_file_formats</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">)</span>		
		
		<span class="c">#print annotated_file_paths</span>
								
	<span class="c">#SETUP COMPLETE. PROCEED TO RUNNING FUNCTION#			</span>
	<span class="c">############################################################################</span>
		
	<span class="c">#the following files will be generated during the program</span>
	<span class="k">global</span> <span class="n">dict_summary</span>
	<span class="k">global</span> <span class="n">codeDict_output_dict</span>
	<span class="c">#global collapsed_output_dict</span>
	<span class="k">global</span> <span class="n">error_file</span>
	<span class="k">global</span> <span class="n">usearch_cluster_file</span>
	<span class="k">global</span> <span class="n">collapsed_output_file</span>
	<span class="k">global</span> <span class="n">summary_file</span>
	<span class="k">global</span> <span class="n">annotation_path</span>	
	<span class="k">global</span> <span class="n">clustered_final_output_file</span>		
		
	
	<span class="c">#stores a file of unique CDR3 pairs. Its a tab delimited file where the first column is the counts for that CDRH3 pair. Remaining columns contain AB information such as CDR3, reocombination type, V,D,J genes, isotypes, average SHM</span>
	<span class="n">collapsed_output_file</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s">&quot;_CDR3_identical_nt_pairs&quot;</span>
	<span class="c">#stores a file of the parsed output from USEARCH clustering. It will store the seed CDR3 sequence and then store all other sequences above 1 read that were clustered into that seed. </span>
	<span class="c">#it is a TAB delimited file </span>
	<span class="n">clustered_output_file</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s">&quot;_CDR3_raw_clustered_nt_pairs_over1read&quot;</span>
	<span class="c">#stores a summary file of all CDR3 clusters created from usearch. It stores the seed sequence, then includes information about this size of the cluster and develops &quot;confidence&quot; and &quot;dominance&quot; terms for each cluster size</span>
	<span class="n">clustered_final_output_file</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s">&quot;_CDR3_clustered_nt_pairs_over1read&quot;</span>
	<span class="c">#this is a JSON file that stores all of the sequences which passed filtering. It stores all the sequence information used during pairing. </span>
	<span class="c">#the key in this file represents the fullcode from a miseq header and the value represents VDJ OR VJ sequence data. </span>
	<span class="n">codeDict_output_dict</span><span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s">&quot;_codeDict&quot;</span>		
	<span class="c">#this is a JSON file created after USEARCH clustering. For each selected cluster it will store the seed as the key and all of the AB information for the CDR3 sequence found in the sequence header submitted to usearch</span>
	<span class="c">#consider removing this file....</span>
	<span class="n">clustered_output_dict</span><span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s">&quot;_clustered_dict&quot;</span>
	<span class="c">#this is a series of JSON strings </span>
	<span class="c">#each line in the file represents either a VH or VL result.</span>
	<span class="c">#when using json.loads(line) on each line, it yeilds a 2 element list. the first element is a dictionary of VDJ data only. The second element is a dictionary of VJ data only.</span>
	<span class="n">dict_summary</span><span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s">&quot;_dict_summary_line_by_line&quot;</span>	
	<span class="c">#this will generate a final summary file describing how many sequences were parsed, how many were sucessful, and how many clusters were formed. it also includes gene usage for H/L</span>
	<span class="n">summary_file</span><span class="o">=</span><span class="n">prefix</span><span class="o">+</span><span class="s">&quot;_Summary.txt&quot;</span>		
	<span class="c">#the input FASTA file submitted to usearch </span>
	<span class="n">usearch_cluster_file</span> <span class="o">=</span> <span class="n">parent_dir</span><span class="o">+</span><span class="s">&quot;clustered_input_file&quot;</span>
	<span class="c">#the output file generated by usearch program </span>
	<span class="n">usearch_output_file</span> <span class="o">=</span> <span class="n">parent_dir</span><span class="o">+</span><span class="s">&#39;clustered_output_file&#39;</span>	
	
	<span class="c">#THIS IS THE ACTUAL CODE FOR RUNNING PAIRING (THIS CALLS ALL THE FUNCTIONS FOR PAIRING)	</span>
	<span class="c">#first create the variables &#39;codeDict&#39; and &#39;collapsedDict&#39; which will store information regarding R1/R2 pairs from the annotation files 	</span>
	<span class="p">[</span><span class="n">total_seqs</span><span class="p">,</span><span class="n">passed_filter</span><span class="p">,</span><span class="n">no_result</span><span class="p">,</span><span class="n">no_cdr3_error</span><span class="p">,</span><span class="n">not_productive</span><span class="p">,</span><span class="n">invalid_chain</span><span class="p">,</span><span class="n">num_unique_cdr3</span><span class="p">,</span><span class="n">num_unique_cdr3_pair_above1</span><span class="p">,</span><span class="n">num_vdj</span><span class="p">,</span><span class="n">num_vj</span><span class="p">]</span> <span class="o">=</span> <span class="n">add_to_dict_memory_safe</span><span class="p">(</span><span class="n">annotated_file_paths</span><span class="p">,</span><span class="n">annotated_file_formats</span><span class="p">,</span><span class="n">required_field_names</span><span class="p">,</span><span class="n">productivity_function</span><span class="p">,</span><span class="n">parent_dir</span><span class="p">,</span><span class="n">analysis_method</span><span class="p">,</span><span class="n">include_vl_analysis</span><span class="p">)</span>		
				
	<span class="n">number_of_clusters</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cluster_cutoff</span><span class="p">:</span>
		<span class="n">usearch_file</span> <span class="o">=</span> <span class="n">usearch_output_file</span><span class="o">+</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;.uc&#39;</span>
	
		<span class="c">#CODEICT CREATED, RUN USEARCH  for all clustered settings</span>
		<span class="k">print</span> <span class="s">&#39;Clustering CDRH3 reads with </span><span class="si">%d%%</span><span class="s"> identity.</span><span class="se">\n</span><span class="s">&#39;</span> <span class="o">%</span><span class="p">(</span><span class="n">c</span><span class="o">*</span><span class="mi">100</span><span class="p">)</span>
		<span class="c">#run USEARCH</span>
		<span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s">&quot;usearch7 -cluster_smallmem {2} -minhsp 10 -minseqlength 10 -usersort -id {0} -centroids c.fa -uc {1}.{0}.uc&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">),</span><span class="n">usearch_output_file</span><span class="p">,</span><span class="n">usearch_cluster_file</span><span class="p">))</span>
		<span class="k">print</span> <span class="s">&#39;USEARCH complete&#39;</span>	
		<span class="k">print</span> <span class="s">&#39;Compiling clustered reads file.</span><span class="se">\n</span><span class="s">&#39;</span>
		<span class="c">#After populate_clusters, clustered_dict_raw[cluster number]=[total count, infoSeed, infoHit 1, infoHit 2, etc]		</span>
		
		<span class="c">#Dict clusters based on %identity of heavy chain. light chain ignored. cluster variable defined at top of script.	</span>
		<span class="n">clustered_dict_raw</span> <span class="o">=</span> <span class="n">populate_clusters</span><span class="p">(</span><span class="n">usearch_file</span><span class="p">,</span><span class="n">delimiter</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>			
			
		<span class="n">clustered_output</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">clustered_output_file</span><span class="o">+</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;.txt&#39;</span><span class="p">,</span><span class="s">&#39;wb&#39;</span><span class="p">)</span>		
		<span class="n">clustered_save</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">clustered_output_dict</span><span class="o">+</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;.txt&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>	
		<span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">clustered_dict_raw</span><span class="o">.</span><span class="n">get</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>			
			<span class="n">CDRH3</span><span class="o">=</span><span class="n">clustered_dict_raw</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">)[</span><span class="mi">3</span><span class="p">]</span>
			<span class="n">clustered_output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%d</span><span class="se">\t</span><span class="si">%s</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span><span class="n">CDRH3</span><span class="p">))</span>
			
			<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">[</span><span class="n">key</span><span class="p">])):</span>
				<span class="n">clustered_output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span><span class="n">clustered_dict_raw</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">))</span>
			<span class="n">clustered_output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>	
			
		<span class="k">print</span> <span class="s">&quot;Saving clustered data.</span><span class="se">\n</span><span class="s">&quot;</span>
		<span class="n">len_clustered</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">)</span>
		<span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">,</span> <span class="n">clustered_save</span><span class="p">)</span>	
		
		<span class="n">clustered_save</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
		<span class="n">clustered_output</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>	
		
		<span class="c">#Prepare clustered file with over 1 read and only the top light chain paired to each heavy chain</span>
		<span class="c">#1st number is rank. </span>
		<span class="c">#2nd number is dominance factor ( LC reads / total number of reads * 100) Good=100, Bad=0</span>
		<span class="c">#3rd number is confidence factor (100 - 100 * (2nd rank LC)/(1st rank LC)) Good=100, Bad=0</span>
		<span class="c">#clustered_dict_raw[cluster number]=[total count, infoSeed, infoHit 1, infoHit 2, etc]	</span>
		<span class="c">#info=(counts,CDRH3_trans,CDRL3_trans,CDRH3,CDRL3,VHgene,DHgene,JHgene,VLgene,JLgene,IgH,IgL,Hmut,Lmut,HLen,LLen)</span>
		<span class="k">print</span><span class="p">(</span><span class="s">&#39;Compiling finalized clustered output file.</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
		<span class="n">Compile_Clusters</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">,</span><span class="n">c</span><span class="p">)</span>		
	
	
		<span class="n">number_of_clusters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">))</span>
	
	<span class="c">#Add in gene histograms to summary file</span>
	<span class="n">gene_hist</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">,</span><span class="n">summary_file</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
	<span class="n">gene_hist</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">,</span><span class="n">summary_file</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
	<span class="n">gene_hist</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">,</span><span class="n">summary_file</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
	<span class="n">gene_hist</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">,</span><span class="n">summary_file</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
	<span class="n">gene_hist</span><span class="p">(</span><span class="n">clustered_dict_raw</span><span class="p">,</span><span class="n">summary_file</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
	<span class="c">#generate a summary file </span>
	<span class="k">if</span> <span class="p">(</span><span class="n">analysis_method</span><span class="o">==</span><span class="s">&#39;IMGT&#39;</span> <span class="ow">and</span> <span class="n">annotated_file_formats</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="s">&#39;IMGT&#39;</span><span class="p">):</span>
		<span class="n">annotated_file_path_summary</span> <span class="o">=</span> <span class="p">[]</span>
		<span class="k">for</span> <span class="n">each_group</span> <span class="ow">in</span> <span class="n">annotated_file_paths</span><span class="p">:</span>
			<span class="n">path</span> <span class="o">=</span> <span class="n">each_group</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)</span>			
			<span class="n">filename</span> <span class="o">=</span> <span class="n">path</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
			<span class="n">path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">each_group</span><span class="p">)</span><span class="c">#+&#39;/&#39;# &#39;/&#39;.join(path[:-1])+&#39;/&#39;</span>
			<span class="n">filename</span> <span class="o">=</span> <span class="n">filename</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;_&#39;</span><span class="p">)</span>
			<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span><span class="o">&gt;=</span><span class="mi">3</span><span class="p">:</span>
				<span class="n">annotated_file_path_summary</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span><span class="s">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filename</span><span class="p">[</span><span class="mi">2</span><span class="p">:])))</span><span class="c"># path+&#39;_&#39;.join(filename[2:]))</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">annotated_file_path_summary</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span><span class="s">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filename</span><span class="p">)))</span><span class="c"># path+&#39;_&#39;.join(filename))</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">annotated_file_path_summary</span> <span class="o">=</span> <span class="n">annotated_file_paths</span>
	<span class="n">WriteSummaryFile</span><span class="p">(</span><span class="n">annotated_file_path_summary</span><span class="p">,</span><span class="n">total_seqs</span><span class="p">,</span><span class="n">passed_filter</span><span class="p">,</span><span class="n">no_result</span><span class="p">,</span><span class="n">no_cdr3_error</span><span class="p">,</span><span class="n">not_productive</span><span class="p">,</span><span class="n">invalid_chain</span><span class="p">,</span><span class="n">num_unique_cdr3</span><span class="p">,</span><span class="n">num_unique_cdr3_pair_above1</span><span class="p">,</span><span class="n">cluster_cutoff</span><span class="p">,</span><span class="n">number_of_clusters</span><span class="p">,</span><span class="n">num_vdj</span><span class="p">,</span><span class="n">num_vj</span><span class="p">)</span>
	
			
	<span class="n">analysis_files_to_report</span> <span class="o">=</span> <span class="p">[</span><span class="n">collapsed_output_file</span><span class="p">,</span><span class="n">summary_file</span><span class="p">]</span>
	<span class="k">for</span> <span class="n">c</span>  <span class="ow">in</span> <span class="n">cluster_cutoff</span><span class="p">:</span>
		<span class="n">analysis_files_to_report</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">usearch_output_file</span><span class="o">+</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;.uc&#39;</span><span class="p">,</span><span class="n">clustered_output_file</span><span class="o">+</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;.txt&#39;</span><span class="p">,</span><span class="n">clustered_final_output_file</span><span class="o">+</span><span class="s">&#39;.&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span><span class="o">+</span><span class="s">&#39;.txt&#39;</span><span class="p">])</span>
	
	
	<span class="k">print</span><span class="p">(</span><span class="s">&#39;All paired clustering complete&#39;</span><span class="p">)</span>
	<span class="k">return</span> <span class="p">{</span><span class="s">&#39;annotation_files&#39;</span><span class="p">:[],</span><span class="s">&#39;analysis_files&#39;</span><span class="p">:</span><span class="n">analysis_files_to_report</span><span class="p">}</span>
</pre></div></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../index.html">IGREP 1.0.0 documentation</a> &raquo;</li>
          <li><a href="index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2015, Costas.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
  </body>
</html>